# 1. CNSs discovery (using poppy as an example)

## 1.1 Mask repeats

```shell
## HN1 means P. somniferum
## YMR means P. rhoeas

BuildDatabase -name HN1 HN1.fasta
RepeatModeler -pa 24 -database HN1
RepeatMasker -pa 24 -dir ./HN1_masker -lib HN1-families.fa HN1.fasta

BuildDatabase -name YMR YMR.fasta
RepeatModeler -pa 24 -database YMR
RepeatMasker -pa 24 -dir ./YMR_masker -lib YMR-families.fa YMR.fasta
```

## 1.2 Prepare some input files

```shell
## Note find_cns related scripts download under URL https://github.com/gturco/find_cns

python /find_cns/pipeline/scripts/gff_to_bed.py --gff HN1.gff3 --fasta /HN1_masker/HN1.fasta.masked  --out /find_cns/pipeline/data/HN1_mask_YMR_mask/HN1_mask
python /find_cns/pipeline/scripts/gff_to_bed.py --gff YMR.gff3 --fasta /YMR_masker/YMR.fasta.masked --out /find_cns/pipeline/data/HN1_mask_YMR_mask/YMR_mask
```



```shell
cp HN1.gff3 YMR.gff3 ./find_cns/pipeline/data/HN1_mask_YMR_mask

awk -F'\t' '{if($3=="CDS") print $0}'  HN1.gff3 > HN1_cds.gff3
awk -F'\t' '{if($3=="CDS") print $0}'  YMR.gff3 > YMR_cds.gff3 

cd /find_cns/pipeline/data/HN1_mask_YMR_mask

bedtools maskfasta -fi HN1_mask.fasta -bed HN1_cds.gff3 -fo HN1_mask.genomic.masked.fasta
bedtools maskfasta -fi YMR_mask.fasta -bed YMR_cds.gff3 -fo YMR_mask.genomic.masked.fasta

cp HN1_mask.bed HN1_mask.all.bed
cp YMR_mask.bed YMR_mask.all.bed

## HN1_mask.all.localdups and YMR_mask.all.localdups are two empty files, without anything. Must have these two named files.

vim HN1_mask.all.localdups
vim YMR_mask.all.localdups

## This file is obtained from your subgenomes.

vim HN1_mask_YMR_mask.pairs.txt

###     
Pso10G24880.0   Prh01G00020.0
Pso10G24870.0   Prh01G00030.0
Pso10G24850.0   Prh01G00050.0
```



## 1.3 Find CNSs

```
MUST NOTE!!!!!

This pipline is just a slightly modified edition for CNS Discovery Pipeline v.3.0.

If you want to use this pipline to find CNSs, you should cite the original authors for this pipline!

Turco, G., Schnable, J. C., Pedersen, B., & Freeling, M. Automated conserved noncoding sequence (CNS) discovery reveals differences in gene content and promoter evolution among grasses. Frontiers in Plant Science, 4, 170.
```





```shell
#!/bin/sh
BLAST_DIR=../cns_pipeline/bin/blast-2.2.26/bin/

ORGA=HN1_mask
ORGB=YMR_mask
QUOTA=2:1
NCPU=`python -c "import multiprocessing; print min(multiprocessing.cpu_count(),18)";`
#### random give two numbers
QDSGID=12345678
SDSGID=12345679

#############################################
# dont edit below here
#############################################
DIR=data/${ORGA}_${ORGB}/


echo checking for unannotated proteins......
##sh coann/co-anno.sh ${ORGA} ${ORGB} $QUOTA $BLAST_DIR


echo finding cns...
python scripts/find_cns.py \
	-q $DIR/${ORGA}.fasta --qbed $DIR/${ORGA}.all.bed \
	-s $DIR/${ORGB}.fasta --sbed $DIR/${ORGB}.all.bed \
        -p $DIR/${ORGA}_${ORGB}.pairs.txt \
        -F T \
        -n 8 \
        --qpad 12000 \
        --spad 12000 \
        --blast_path ${BLAST_DIR}\bl2seq \
        --pair_fmt pair > $DIR/${ORGA}_${ORGB}.cns.txt 
        ### if using blast+  --blast_path ${BLAST_DIR}\legacy_blast.pl \

status=$?
if [ $status -ne 0 ]; then
  echo "Error runnining find cns"
  exit
fi


python scripts/localdup.py \
       -q $DIR/${ORGA}.fasta --qbed $DIR/${ORGA}.all.bed \
       -s $DIR/${ORGB}.fasta --sbed $DIR/${ORGB}.all.bed \
        -p $DIR/${ORGA}_${ORGB}.pairs.txt \
       --cns_file $DIR/${ORGA}_${ORGB}.cns.txt \
        -F T \
        -n 8 \
        --qpad 12000 \
        --spad 12000 \
        --blast_path ${BLAST_DIR}\bl2seq \
       --pair_fmt pair \
       --qdups $DIR/${ORGA}.all.localdups \
       --sdups $DIR/${ORGB}.all.localdups
        ### if using blast+  --blast_path ${BLAST_DIR}\legacy_blast.pl \
status=$?
if [ $status -ne 0 ]; then
  echo "Error runnining finding localdups"
  exit
fi


python scripts/post_processing/cns_to_fasta.py \
                -c $DIR/${ORGA}_${ORGB}.cns.txt.local \
                --qfasta $DIR/${ORGA}.genomic.masked.fasta \
                --sfasta $DIR/${ORGB}.genomic.masked.fasta \
                --qorg ${ORGA} \
                --sorg ${ORGB} \
                --min_len=18 \
                > $DIR/${ORGA}_${ORGB}.cns_test.fasta

status=$?
if [ $status -ne 0 ]; then
  echo "Error runnining cns2fasta"
  exit
fi


echo removing cns that have hits in arabidopsis as rna or protein

## we modify these parts and include protein annotations for HN1 and YMR

bblast.py -b ${BLAST_DIR}/blastall -p blastx -d data/at_protein.fasta -i $DIR/${ORGA}_${ORGB}.cns_test.fasta -e 0.01 -m 8 -a ${NCPU} -o $DIR/at_protein.blast
bblast.py -b ${BLAST_DIR}/blastall -p blastx -d data/HN1_YMR.pep -i $DIR/${ORGA}_${ORGB}.cns_test.fasta -e 0.01 -m 8 -a ${NCPU} -o $DIR/HN1_YMR_protein.blast

status=$?
if [ $status -ne 0 ]; then
  echo "Error running bblast"
  exit
fi


python scripts/post_processing/find_exons.py \
                 --cns $DIR/${ORGA}_${ORGB}.cns.txt.local \
                 -o $DIR \
                 $DIR/at_protein.blast $DIR/HN1_YMR_protein.blast

status=$?
if [ $status -ne 0 ]; then
  echo "Error running find exons"
  exit
fi



python scripts/post_processing/find_rna.py \
	    --blastpath $BLAST_DIR/blastall \
      -g data/thaliana_v10.gff \
         -f data/thaliana_v10.fasta \
         -b $DIR/${ORGA}_${ORGB}_cns_vs_at_rnas.blast \
     	 -q ${ORGA}  \
      	 -s ${ORGB} \
   	 -o $DIR \
   	 -d data/thaliana_v10.description


status=$?
if [ $status -ne 0 ]; then
  echo "Error running find rna"
  exit
fi


cp $DIR/${ORGA}_${ORGB}.raw.filtered.local  $DIR/${ORGA}_${ORGB}.raw2.filtered.local

python scripts/post_processing/shuffle_protein_cns.py \
   --qbed $DIR/${ORGA}.all.nolocaldups.bed.local \
   --sbed $DIR/${ORGB}.all.nolocaldups.bed.local \
   --cns  $DIR/${ORGA}_${ORGB}.cns.txt.local \
   --paralogy  $DIR/${ORGA}_${ORGB}.raw.filtered.local \
   --orthology $DIR/${ORGA}_${ORGB}.raw2.filtered.local \
   --pairs $DIR/${ORGA}_${ORGB}.pairs.txt.local \ 
#creates: $DIR/${ORGA}_${ORGB}.quota.with_new.orthology

status=$?
if [ $status -ne 0 ]; then
  echo "Error running shuffle proteins"
  exit
fi



python scripts/post_processing/assign.py \
     --qbed $DIR/${ORGA}.all.nolocaldups.bed.with_new.local \
     --sbed $DIR/${ORGB}.all.nolocaldups.bed.with_new.local \
     --cns $DIR/${ORGA}_${ORGB}.cns.txt.real.local \
     --pairs $DIR/${ORGA}_${ORGB}.pairs.txt.local \
     --qdsgid $QDSGID \
     --sdsgid $SDSGID \
     --qpad 15000 \
     --spad 15000 \
     --pair_fmt pair > $DIR/${ORGA}_${ORGB}.cns.assigned.csv.local


status=$?
if [ $status -ne 0 ]; then
  echo "Error running assigning cns to genes"
  exit
fi



python scripts/post_processing/cns_to_fasta.py \
                -c $DIR/${ORGA}_${ORGB}.cns.txt.real.local \
                --qfasta $DIR/${ORGA}.genomic.masked.fasta \
                --sfasta $DIR/${ORGB}.genomic.masked.fasta \
                --qorg ${ORGA} \
                --sorg ${ORGB} \
                > $DIR/${ORGA}_${ORGB}.cns.fasta


status=$?
if [ $status -ne 0 ]; then
  echo "Error running cns to fasta"
  exit
fi



### creates genelist and cnslist
sh list.sh ${ORGA} ${ORGB} $QUOTA $QDSGID $SDSGID


```

