# 1. Identification of two subgenomes in P. somniferum

## 1.1 Blastp

```shell
## HN1 means P. somniferum
## YMR means P. rhoeas
blastp -query HN1.pep -db HN1 -evalue 1e-5 -outfmt 6 -num_threads 48 -max_target_seqs 11 -out HN1_11.blast
blastp -query YMR.pep -db YMR -evalue 1e-5 -outfmt 6 -num_threads 48 -max_target_seqs 11  -out YMR_11.blast
blastp -query HN1.pep -db YMR -evalue 1e-5  -outfmt 6 -num_threads 48 -max_target_seqs 10 -out HN1_YMR_10.blast
```



## 1.2 Filter Blastp results

```R
rm(list = ls())
gene_blast <- read.table("HN1_11.blast",
                       sep="\t",stringsAsFactors =FALSE)
gene_blast <- gene_blast[gene_blast[,12] >= 100,]
write.table(gene_blast,"Ps.blast"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```

```R
rm(list = ls())
gene_blast <- read.table("YMR_11.blast",
                       sep="\t",stringsAsFactors =FALSE)
gene_blast <- gene_blast[gene_blast[,12] >= 100,]
write.table(gene_blast,"Pr.blast"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```

## 1.3 Identify Tandem duplicates

```R
##make specific gff files for MCScanX
rm(list = ls())
gene_bed <- read.table("Papaver_somniferum.gene.bed",
                       sep="\t",stringsAsFactors =FALSE)

gene_bed[1:5,]
'''
    V1     V2     V3            V4 V5
1 chr1  59647  63768 Pso01G00010.0  +
2 chr1  62504  73904 Pso01G00020.0  -
3 chr1 142842 146495 Pso01G00030.0  +
4 chr1 145932 151557 Pso01G00040.0  -
5 chr1 151746 155004 Pso01G00050.0  +
'''
gene_bed[,1] <- paste0("Ps",gene_bed[,1])
gene_bed <- gene_bed[,c(1,4,2,3)]

write.table(gene_bed,"/subgenome/TD/Ps/Ps.gff"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

rm(list = ls())
gene_bed <- read.table("Papaver_rhoeas.gene.bed",
                       sep="\t",stringsAsFactors =FALSE)
gene_bed[,1] <- paste0("Pr",gene_bed[,1])
gene_bed <- gene_bed[,c(1,4,2,3)]

write.table(gene_bed,"/subgenome/TD/Pr/Pr.gff"
            ,sep = "\t",col.names = F,row.names = F,quote = F)
```



```shell
## Ps means P. somniferum
## Pr means P. rhoeas

./MCScanX ../../subgenome/TD/Ps/Ps
./MCScanX ../../subgenome/TD/Pr/Pr
```



## 1.4 Filter Blastp results (Remove TDs)

```R
rm(list = ls())
gene_blast <- read.table("HN1_YMR_10.blast",
                       sep="\t",stringsAsFactors =FALSE)
HN1_tandem <- read.table("/subgenome/TD/Ps/Ps.tandem",
                         sep=",",stringsAsFactors =FALSE)
HN1_tandem_list <- as.character(unique(c(HN1_tandem[,1],HN1_tandem[,2])))


YMR_tandem <- read.table("/subgenome/TD/Pr/Pr.tandem",
                         sep=",",stringsAsFactors =FALSE)
YMR_tandem_list <- as.character(unique(c(YMR_tandem[,1],YMR_tandem[,2])))

gene_blast<- gene_blast[!gene_blast[,1] %in% HN1_tandem_list,]
gene_blast<- gene_blast[!gene_blast[,2] %in% YMR_tandem_list,]

write.table(gene_blast,"/subgenome/HN1_YMR_wgdi/HN1_YMR_10_rdup.blast"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```

## 1.5 Identify collinearity for P. somniferum

```shell
[dotplot]
blast = HN1_HN1_rdup.blast
gff1 =  HN1.gff
gff2 =  HN1.gff
lens1 = HN1_chr_only.lens
lens2 = HN1_chr_only.lens
genome1_name =  Papaver_somniferum
genome2_name =  Papaver_somniferum
multiple  = 2
score = 100
evalue = 1e-5
repeat_number = 5
position = order
blast_reverse = false
ancestor_left = none
ancestor_top = none
markersize = 0.5
figsize = 14.21,14.21
savefig = HN1_HN1_dotplot.png

[collinearity]
gff1 =  HN1.gff
gff2 =  HN1.gff
lens1 = HN1.lens
lens2 = HN1.lens
blast = HN1_HN1_rdup.blast
blast_reverse = false
multiple  = 2
process = 24
evalue = 1e-5
score = 100
grading = 50,40,25
mg = 40,40
pvalue = 0.2
repeat_number = 5
positon = order
savefile = HN1_HN1.collinearity.txt

[ks]
cds_file = 	HN1.cds.fa
pep_file = 	HN1.pep
align_software = muscle
pairs_file = HN1_HN1.collinearity.txt
ks_file = HN1_HN1.collinearity.ks


[blockinfo]
blast = HN1_HN1_rdup.blast
gff1 =  HN1.gff
gff2 =  HN1.gff
lens1 = HN1.lens
lens2 = HN1.lens
collinearity = HN1_HN1.collinearity.txt
score = 100
evalue = 1e-5
repeat_number = 5
position = order
ks = HN1_HN1.collinearity.ks
ks_col = ks_YN00
savefile = HN1_HN1_block_information.csv


[blockks]
lens1 = HN1_chr_only.lens
lens2 = HN1_chr_only.lens
genome1_name =  Papaver_somniferum
genome2_name =  Papaver_somniferum
blockinfo = HN1_HN1_block_information.csv
pvalue = 0.2
tandem = true
tandem_length = 20
markersize = 1
area = -1,3
block_length =  10
figsize = 14.21,14.21
savefig = HN1_HN1_ks_dotplot.pdf
```

  

```R
## Filter the diagonal blocks

all_block <- read.csv("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/HN1_HN1_block_information.csv",stringsAsFactors = F)
lens <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/HN1.lens",sep = "\t",stringsAsFactors = F)
result_data <- rbind()

for(i in 1:1132){
  all_block_sub <- all_block[i,]
  if(all_block_sub[1,2] != "chr1"){
    chr_pre <- lens[1:(as.numeric(rownames(lens[lens[,1] == all_block_sub[1,2],]))-1),][,1]
  }else{
    chr_pre <- "escape"
  }
  
  if(!all_block_sub[1,3] %in% chr_pre){
    if(all_block_sub[1,3] != all_block_sub[1,2]){
      result_data <- rbind(result_data,all_block_sub)
    }else if(max(all_block_sub[1,4],all_block_sub[1,5]) <=  min(all_block_sub[1,6],all_block_sub[1,7])){
      result_data <- rbind(result_data,all_block_sub)
    }
  }
    
}
write.table(result_data,"D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/HN1_HN1_block_information_filter.csv"
            ,sep = ",",col.names = T,row.names = F,quote = F)

```



```shell
[blockks]
lens1 = HN1_chr_only.lens
lens2 = HN1_chr_only.lens
genome1_name =  Papaver_somniferum
genome2_name =  Papaver_somniferum
blockinfo = HN1_HN1_block_information_filter.csv
pvalue = 0.2
tandem = true
tandem_length = 20
markersize = 1
area = -1,3
block_length =  10
figsize = 14.21,14.21
savefig = HN1_HN1_ks_dotplot_filter.pdf

[kspeaks]
blockinfo = HN1_HN1_block_information_filter.csv 
pvalue = 0.2
tandem = true
block_length = 10
ks_area = 0,10
multiple  = 2
homo = 0.5,1
fontsize = 9
area = 0,3
figsize = 10,10
savefig = HN1_HN1.ks_median.distri.pdf
savefile = HN1_HN1.ks_median.distri.csv


```



## 1.6 preliminary identify WGD pairs and singletons

```R
HN1_gene_list_block1<-
  function(df,gff_HN1){
    
    gene_HN1 <- rbind()
    for (i in 1:dim(df)[1]) {
      syn_block1 <-  df[i,c("block1")]
      chr <- df[i,2]
      gene_order <- unlist(
        lapply(syn_block1, strsplit, split="_", fixed=TRUE, useBytes=TRUE)
      )
      
      gene_order <- unique(as.numeric(gene_order))
      gff_1 <- gff_HN1[gff_HN1[,1] == chr,]
      
      syn_block1_gene <- gff_1[ gff_1$V6 %in% gene_order, "V2"]
      syn_block1_gene <- as.data.frame(syn_block1_gene)
      syn_block1_gene[,2] <- paste0(df[i,1],"_","block1")
      colnames(syn_block1_gene) <- c("gene","block")
      gene_HN1 <- rbind(gene_HN1,syn_block1_gene)
    }
    return(gene_HN1)
  }

HN1_gene_list_block2<-
  function(df,gff_HN1){
    
    gene_HN1 <- rbind()
    for (i in 1:dim(df)[1]) {
      syn_block1 <-  df[i,c("block2")]
      chr <- df[i,3]
      gene_order <- unlist(
        lapply(syn_block1, strsplit, split="_", fixed=TRUE, useBytes=TRUE)
      )
      
      gene_order <- unique(as.numeric(gene_order))
      gff_1 <- gff_HN1[gff_HN1[,1] == chr,]
      
      syn_block1_gene <- gff_1[ gff_1$V6 %in% gene_order, "V2"]
      syn_block1_gene <- as.data.frame(syn_block1_gene)
      syn_block1_gene[,2] <- paste0(df[i,1],"_","block2")
      colnames(syn_block1_gene) <- c("gene","block")
      gene_HN1 <- rbind(gene_HN1,syn_block1_gene)
    }
    return(gene_HN1)
  }

```



```R
all_block <- read.csv("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/HN1_HN1_block_information.csv",stringsAsFactors = F)

gff_HN1 <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/HN1.gff",stringsAsFactors = FALSE)



all_block_1 <- HN1_gene_list_block1(all_block,gff_HN1)
all_block_2 <- HN1_gene_list_block2(all_block,gff_HN1)


all_wgd_genes <- as.character(unique(c(as.character(all_block_1[,1]),as.character(all_block_2[,1]))))

HN1_tandem <- read.table("D:/002cns_evolution/subgenome/TD/Ps/Ps.tandem",
                         sep=",",stringsAsFactors =FALSE)
HN1_tandem_list <- as.character(unique(c(HN1_tandem[,1],HN1_tandem[,2])))

singleton <- gff_HN1[!gff_HN1[,2] %in% all_wgd_genes,]
singleton <- singleton[!singleton[,2] %in% HN1_tandem_list,]
write.table(singleton,"D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/singleton.xls"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```

```R
all_block <- read.csv("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/HN1_HN1.ks_median.distri_false.csv",stringsAsFactors = F)

all_block <- all_block[all_block$ks_median < 0.2,]
gff_HN1 <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/HN1.gff",stringsAsFactors = FALSE)



all_block_1 <- HN1_gene_list_block1(all_block,gff_HN1)
all_block_2 <- HN1_gene_list_block2(all_block,gff_HN1)

WGD_pairs_genes <- as.character(unique(c(as.character(all_block_1[,1]),as.character(all_block_2[,1]))))
WGD_pairs_genes_block <- unique(rbind(all_block_1,all_block_2))

write.table(WGD_pairs_genes,"D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/WGD_pairs_genes.txt"
            ,sep = "\t",col.names = F,row.names = F,quote = F)
write.table(WGD_pairs_genes_block,"D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/WGD_pairs_genes_block.xls"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```





## 1.7 Identify collinearity between P. somniferum and P. rhoeas

```R
rm(list = ls())
## filter blastp results
gene_blast <- read.table("HN1_YMR_10.blast",
                         sep="\t",stringsAsFactors =FALSE)

HN1_tandem <- read.table("Ps.tandem",
                         sep=",",stringsAsFactors =FALSE)
HN1_tandem_list <- as.character(unique(c(HN1_tandem[,1],HN1_tandem[,2])))


YMR_tandem <- read.table("Pr.tandem",
                         sep=",",stringsAsFactors =FALSE)
YMR_tandem_list <- as.character(unique(c(YMR_tandem[,1],YMR_tandem[,2])))

gene_blast<- gene_blast[!gene_blast[,1] %in% HN1_tandem_list,]
gene_blast<- gene_blast[!gene_blast[,2] %in% YMR_tandem_list,]

singleton <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/singleton.xls",
                        sep="\t",stringsAsFactors =FALSE)

WGD_pairs <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/WGD_pairs_genes.txt",
                        sep="\t",stringsAsFactors =FALSE)
gene_list <- as.character(unique(c(singleton[,2],WGD_pairs[,1])))

gene_blast<- gene_blast[gene_blast[,1] %in% gene_list,]


write.table(gene_blast,"D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi/HN1_YMR_10_rdup.blast"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```

```shell
[dotplot]
blast = HN1_YMR_10_rdup.blast
gff1 =  HN1.gff
gff2 =  YMR.gff
lens1 = HN1_chr_only.lens
lens2 = YMR_chr_only.lens
genome1_name =  Papaver_somniferum
genome2_name =  Papaver_rhoeas
multiple  = 2
score = 100
evalue = 1e-5
repeat_number = 10
position = order
blast_reverse = false
ancestor_left = none
ancestor_top = none
markersize = 0.5
figsize = 12,14.21
savefig = HN1_YMR_dotplot.pdf

[collinearity]
gff1 =  HN1.gff
gff2 =  YMR.gff
lens1 = HN1.lens
lens2 = YMR.lens
blast = HN1_YMR_10_rdup.blast
blast_reverse = false
multiple  = 2
process = 24
evalue = 1e-5
score = 100
grading = 50,40,25
mg = 40,40
pvalue = 0.2
repeat_number = 5
positon = order
savefile = HN1_YMR.collinearity.txt

[ks]
cds_file = 	HN1_YMR.cds.fa
pep_file = 	HN1_YMR.pep
align_software = muscle
pairs_file = HN1_YMR.collinearity.txt
ks_file = HN1_YMR.collinearity.ks

[blockinfo]
blast = HN1_YMR_10_rdup.blast
gff1 =  HN1.gff
gff2 =  YMR.gff
lens1 = HN1.lens
lens2 = YMR.lens
collinearity = HN1_YMR.collinearity.txt
score = 100
evalue = 1e-5
repeat_number = 5
position = order
ks = HN1_YMR.collinearity.ks
ks_col = ks_YN00
savefile = HN1_YMR_block_information.csv

[blockks]
lens1 = HN1_chr_only.lens
lens2 = YMR_chr_only.lens
genome1_name =  Papaver_somniferum
genome2_name =  Papaver_rhoeas
blockinfo = HN1_YMR_block_information.csv
pvalue = 0.2
tandem = true
tandem_length = 20
markersize = 1
area = -1,3
block_length =  10
figsize = 12,14.21
savefig = HN1_YMR_ks_dotplot.pdf

[kspeaks]
blockinfo = HN1_YMR_block_information.csv 
pvalue = 0.2
tandem = true
block_length = 10
ks_area = 0,10
multiple  = 2
homo = 0.5,1
fontsize = 9
area = 0,3
figsize = 10,6.18
savefig = HN1_YMR.ks_median.distri.pdf
savefile = HN1_YMR.ks_median.distri.csv

```





## 1.8 Reconstruct subgenomes of P. somniferum based on k-mer

### 1.8.1 SubPhaser to identify specific k-mer

```shell
git clone https://github.com/zhangrengang/SubPhaser
unzip SubPhaser-master.zip 
conda env create -f SubPhaser.yaml
conda activate SubPhaser
python setup.py install
```



```shell
vim HN1.config

chr1	chr6
chr2
chr3
chr4
chr5
chr7
chr8
chr11	chr9
chr10
1|unplaced-scaffold_1
10|unplaced-scaffold_10
14|unplaced-scaffold_14
19|unplaced-scaffold_19
2|unplaced-scaffold_2
24|unplaced-scaffold_24
28|unplaced-scaffold_28
5|unplaced-scaffold_5
7|unplaced-scaffold_7
9|unplaced-scaffold_9
```

```shell
subphaser -i \HN1_unmask.fasta -c ./HN1.config
```

### 1.8.2 Reconstruct 

```R


subphaser_result <- read.table("D:/002cns_evolution_1st_revision/001subgenome/subphaserHN1Result_all/k15_q200_f2.bin.enrich",
                               sep = "\t",stringsAsFactors = F)
colnames(subphaser_result) <- c("chrom",	"start"	,"end",	
                                "subgenome",	"p_value",	"counts",	"ratios",
                                "enrich",	"pvals",	"potential_exchange",	"p_corrected")
gff_HN1 <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/HN1.gff",stringsAsFactors = FALSE)

gff_HN1[,8] <- NA

for (i in 1:55136) {
  subphaser_result_sub <- subphaser_result[subphaser_result[,1] == gff_HN1[i,1],]
  subphaser_result_sub <- subphaser_result_sub[subphaser_result_sub$start <= gff_HN1[i,3] & subphaser_result_sub$end > gff_HN1[i,3],]
  if(dim(subphaser_result_sub)[1] != 0){
    gff_HN1[i,8] <- unique(subphaser_result_sub$subgenome)
  }else{
    gff_HN1[i,8] <- "None"
  }
  
}


gff_HN1_table <- as.data.frame(table(gff_HN1[,8]))
write.table(gff_HN1,"D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/subphaser_gene.gff"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```

```R
HN1_gene_list<-
  function(df){
    gff_HN1 <- read.table("D:/002cns_evolution/subgenome/TD/HN1/HN1.gff",stringsAsFactors = FALSE)
    
    gene_HN1 <- rbind()
    for (i in 1:dim(df)[1]) {
      syn_block1 <-  df[i,c("block1")]
      chr <- df[i,2]
      gene_order <- unlist(
        lapply(syn_block1, strsplit, split="_", fixed=TRUE, useBytes=TRUE)
      )
      
      gene_order <- unique(as.numeric(gene_order))
      gff_1 <- gff_HN1[gff_HN1[,1] == chr,]
      
      syn_block1_gene <- gff_1[ gff_1$V6 %in% gene_order, "V2"]
      syn_block1_gene <- as.data.frame(syn_block1_gene)
      syn_block1_gene[,2] <- df[i,1]
      colnames(syn_block1_gene) <- c("gene","block")
      gene_HN1 <- rbind(gene_HN1,syn_block1_gene)
    }
    return(as.character(unique(gene_HN1[,1])))
  }

```

```R
## the first edition
all_block_hn1_ymr <- read.csv("D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi/HN1_YMR.ks_median.distri.csv",stringsAsFactors = F)
all_block_hn1_ymr <- all_block_hn1_ymr[all_block_hn1_ymr$ks_median < 0.5,]

subphaser_result <- read.table("D:/002cns_evolution_1st_revision/001subgenome/subphaserHN1Result_all/k15_q200_f2.bin.enrich",
                               sep = "\t",stringsAsFactors = F)
colnames(subphaser_result) <- c("chrom",	"start"	,"end",	
                                "subgenome",	"p_value",	"counts",	"ratios",
                                "enrich",	"pvals",	"potential_exchange",	"p_corrected")
gff_HN1 <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/subphaser_gene.gff",stringsAsFactors = FALSE)

result_data <- rbind()
for (i in 1:444) {
  HN1gene <- HN1_gene_list(all_block_hn1_ymr[i,])
  gff_HN1_sub <- gff_HN1[gff_HN1[,2] %in% HN1gene, ]
  gff_HN1_sub_table <- as.data.frame(table(gff_HN1_sub[,8]))
  gff_HN1_sub_table[,3] <- all_block_hn1_ymr[i,1]
  result_data <- rbind(result_data,gff_HN1_sub_table)
  
}

result_data_table <- as.data.frame(table(result_data[,3]))


for (i in 1:444) {
  print(i)
  result_data_sub <- result_data[result_data[,3] ==all_block_hn1_ymr[i,1], ]
  class_index <- as.character(result_data_sub[result_data_sub[,2] == max(result_data_sub[,2]),][1,1])
  if(class_index == "SG1"){
    all_block_hn1_ymr[i,22:23] <- 1
  }else if(class_index == "SG2"){
    all_block_hn1_ymr[i,22:23] <- 2
  }else if(class_index == "None"){
    all_block_hn1_ymr[i,22:23] <- 3
  }
  
}

write.table(all_block_hn1_ymr,"D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/HN1_YMR_first_assemble.csv"
            ,sep = ",",col.names = T,row.names = F,quote = F)

```

```R
## the second edition with some manual corrections

all_block_hn1_ymr <- read.csv("D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/HN1_YMR_first_assemble.csv",stringsAsFactors = F)

all_block_hn1_ymr[all_block_hn1_ymr[,1] == 9,22:23] <- 1

all_block_hn1_ymr[all_block_hn1_ymr[,1] == 263,22:23] <- 1

all_block_hn1_ymr[all_block_hn1_ymr[,2] == "chr3" & all_block_hn1_ymr[,3] == "chr4",22:23] <- 1

all_block_hn1_ymr[all_block_hn1_ymr[,2] == "chr3" & all_block_hn1_ymr[,3] == "chr6",22:23] <- 2

all_block_hn1_ymr[all_block_hn1_ymr[,2] == "chr3" & all_block_hn1_ymr[,3] == "chr7",22:23] <- 1

all_block_hn1_ymr[all_block_hn1_ymr[,2] == "chr4" & all_block_hn1_ymr[,3] == "chr3",22:23] <- 1

all_block_hn1_ymr[all_block_hn1_ymr[,2] == "chr5" & all_block_hn1_ymr[,3] == "chr4",22:23] <- 2

all_block_hn1_ymr[all_block_hn1_ymr[,2] == "chr9" & all_block_hn1_ymr[,3] == "chr2",22:23] <- 1


all_block_hn1_ymr[all_block_hn1_ymr[,1] == 102,22:23] <- 1
all_block_hn1_ymr[all_block_hn1_ymr[,1] == 98,22:23] <- 1


all_block_hn1_ymr[all_block_hn1_ymr[,2] == "chr11",22:23] <- 2

all_block_hn1_ymr[all_block_hn1_ymr[,2] == "unplaced-scaffold_1" & all_block_hn1_ymr[,3] == "chr1",22:23] <- 2

all_block_hn1_ymr[all_block_hn1_ymr[,2] == "unplaced-scaffold_1" & all_block_hn1_ymr[,3] == "chr3",22:23] <- 1

all_block_hn1_ymr[all_block_hn1_ymr[,2] == "unplaced-scaffold_1" & all_block_hn1_ymr[,3] == "chr4",22:23] <- 1


all_block_hn1_ymr[all_block_hn1_ymr[,2] == "unplaced-scaffold_2" & all_block_hn1_ymr[,3] == "chr2",22:23] <- 1


all_block_hn1_ymr[all_block_hn1_ymr[,2] == "unplaced-scaffold_5" & all_block_hn1_ymr[,3] == "chr1",22:23] <- 2

write.table(all_block_hn1_ymr,"D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/HN1_YMR_2nd_assemble.csv"
            ,sep = ",",col.names = T,row.names = F,quote = F)

```



```shell
[alignment]
gff1 =  ../HN1_YMR_wgdi/HN1.gff
gff2 =  ../HN1_YMR_wgdi/YMR.gff
lens1 = repeat5_HN1.lens
lens2 = ../HN1_YMR_wgdi/YMR_chr_only.lens
genome1_name =  ../HN1_YMR_wgdi/Papaver_somniferum
genome2_name =  ../HN1_YMR_wgdi/Papaver_rhoeas
markersize = 0.5
ks_area = 0,3
position = order
colors = red,blue,green
figsize = 12,15
savefile = HN1_YMR_2nd_assemble_alignment.csv
savefig= HN1_YMR_2nd_assemble_alignment.pdf
blockinfo = HN1_YMR_2nd_assemble.csv
blockinfo_reverse = false
classid =  class1


wgdi -a HN1_YMR_wgdi.conf
```

```R
alignment <- read.csv("D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/HN1_YMR_2nd_assemble_alignment.csv",stringsAsFactors = F,
                      header = F)

alignment_sub1 <- alignment[alignment[,2] != "" & alignment[,2] != ".",]
alignment_sub1 <- alignment_sub1[,-3]
colnames(alignment_sub1) <- c("HN1","YMR")
alignment_sub2 <- alignment[alignment[,3] != "" & alignment[,3] != ".",]
alignment_sub2 <- alignment_sub2[,-2]
colnames(alignment_sub2) <- c("HN1","YMR")

a1_dup <- alignment_sub1[duplicated(alignment_sub1[,2]),]
a2_dup <-alignment_sub2[duplicated(alignment_sub2[,2]),]
dup <- unique(c(a1_dup[,2],a2_dup[,2]))

alignment_sub1 <- alignment_sub1[!alignment_sub1[,2] %in% dup,]
alignment_sub2 <- alignment_sub2[!alignment_sub2[,2] %in% dup,]

alignment_sub_all <- intersect(alignment_sub1[,1],alignment_sub2[,1])

alignment_sub1 <- alignment_sub1[!alignment_sub1[,1] %in% alignment_sub_all,]
alignment_sub2 <- alignment_sub2[!alignment_sub2[,1] %in% alignment_sub_all,]

YMR_gff <- read.table("D:/002cns_evolution/subgenome/HN1_YMR_wgdi/YMR.gff",stringsAsFactors = F)
YMR_gff <- YMR_gff[YMR_gff$V1 %in% paste0("chr",1:7),]
YMR_gff <- YMR_gff[,c(1:4)]
YMR_gff <- YMR_gff[,c(2:4,1)]
YMR_gff[,c(2,3)] <- ""
colnames(YMR_gff) <- c("YMR_gene","subgenome1","subgenome2","YMR_chr")

for (i in 1:dim(alignment_sub1)[1]) {
  YMR_gff[YMR_gff[,1] == alignment_sub1[i,]$YMR,]$subgenome1 <- alignment_sub1[i,]$HN1
}

for (i in 1:dim(alignment_sub2)[1]) {
  YMR_gff[YMR_gff[,1] == alignment_sub2[i,]$YMR,]$subgenome2 <- alignment_sub2[i,]$HN1
}

statics_result <- as.data.frame(matrix(NA,7,5))
colnames(statics_result) <- c("chr","subgenom1_unique","subgenome2_unique","overlap","p_value")
for (chr in 1:7) {
  chr_i <- paste0("chr",chr)
  YMR_gff_sub <- YMR_gff[YMR_gff$YMR_chr == chr_i,]
  subgenome1_length <- as.data.frame(table(YMR_gff_sub$subgenome1))
  subgenome1_length <- subgenome1_length[subgenome1_length[,1] != "",]
  subgenome1_length <- YMR_gff_sub[YMR_gff_sub$subgenome1 %in% subgenome1_length[,1],]
  
  subgenome2_length <- as.data.frame(table(YMR_gff_sub$subgenome2))
  subgenome2_length <- subgenome2_length[subgenome2_length[,1] != "",]
  subgenome2_length <- YMR_gff_sub[YMR_gff_sub$subgenome2 %in% subgenome2_length[,1],]
  
  statics_result[chr,1] <- chr_i
  statics_result[chr,2] <- length(setdiff(subgenome1_length[,1],intersect(subgenome1_length[,1],subgenome2_length[,1])))
  statics_result[chr,3] <- length(setdiff(subgenome2_length[,1],intersect(subgenome1_length[,1],subgenome2_length[,1])))
  statics_result[chr,4] <- length(intersect(subgenome1_length[,1],subgenome2_length[,1]))
  statics_result[chr,5] <- fisher.test(matrix(c(statics_result[chr,2],statics_result[chr,4],statics_result[chr,3],statics_result[chr,4]),ncol = 2))$p.value
}


write.csv(statics_result,"D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/HN1_YMR_2nd_assemble_statics_subgenome.csv")
write.table(YMR_gff,"D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/HN1_YMR_2nd_assemble_alignment_final_chr.csv",
            sep = ",",col.names = F,row.names = F,quote = F)

```

## 1.9 Reconstruct subgenomes of P. somniferum based on retained differences

```R
HN1_gene_list_block1<-
  function(df,gff_HN1){
    
    gene_HN1 <- rbind()
    for (i in 1:dim(df)[1]) {
      syn_block1 <-  df[i,c("block1")]
      chr <- df[i,2]
      gene_order <- unlist(
        lapply(syn_block1, strsplit, split="_", fixed=TRUE, useBytes=TRUE)
      )
      
      gene_order <- unique(as.numeric(gene_order))
      gff_1 <- gff_HN1[gff_HN1[,1] == chr,]
      
      syn_block1_gene <- gff_1[ gff_1$V6 %in% gene_order, "V2"]
      syn_block1_gene <- as.data.frame(syn_block1_gene)
      syn_block1_gene[,2] <- paste0(df[i,1],"_","block1")
      colnames(syn_block1_gene) <- c("gene","block")
      gene_HN1 <- rbind(gene_HN1,syn_block1_gene)
    }
    return(gene_HN1)
  }

HN1_gene_list_block2<-
  function(df,gff_HN1){
    
    gene_HN1 <- rbind()
    for (i in 1:dim(df)[1]) {
      syn_block1 <-  df[i,c("block2")]
      chr <- df[i,3]
      gene_order <- unlist(
        lapply(syn_block1, strsplit, split="_", fixed=TRUE, useBytes=TRUE)
      )
      
      gene_order <- unique(as.numeric(gene_order))
      gff_1 <- gff_HN1[gff_HN1[,1] == chr,]
      
      syn_block1_gene <- gff_1[ gff_1$V6 %in% gene_order, "V2"]
      syn_block1_gene <- as.data.frame(syn_block1_gene)
      syn_block1_gene[,2] <- paste0(df[i,1],"_","block2")
      colnames(syn_block1_gene) <- c("gene","block")
      gene_HN1 <- rbind(gene_HN1,syn_block1_gene)
    }
    return(gene_HN1)
  }
  
## to enlarge the duplicated blocks to contain some singletons
block1_count <- function(singleton,all_block,i,number_minus){
  
  singleton_1 <- singleton[singleton[,1] == all_block[i,2],]
  
  singleton_1 <- singleton_1[singleton_1[,6] >= min(all_block[i,4],all_block[i,5])-number_minus &
                               singleton_1[,6] <= max(all_block[i,4],all_block[i,5])+number_minus ,]
  
  
  all_block_1 <- HN1_gene_list_block1(all_block[i,],gff_HN1)
  
  block_1_gene <- c(as.character(singleton_1[,2]),as.character(all_block_1[,1]))
  
  return(block_1_gene)
}

block2_count <- function(singleton,all_block,i,number_minus){
  
  singleton_2 <- singleton[singleton[,1] == all_block[i,3],]
  
  singleton_2 <- singleton_2[singleton_2[,6] >= min(all_block[i,6],all_block[i,7])-number_minus &
                               singleton_2[,6] <= max(all_block[i,6],all_block[i,7])+number_minus ,]
  
  all_block_2 <- HN1_gene_list_block2(all_block[i,],gff_HN1)

  block_2_gene <- c(as.character(singleton_2[,2]),as.character(all_block_2[,1]))
  
  return(block_2_gene)
}
```

```R
all_block <- read.csv("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/HN1_HN1.ks_median.distri_false.csv",stringsAsFactors = F)

all_block <- all_block[all_block$ks_median < 0.2,]


singleton <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/singleton.xls",sep = "\t",
                        stringsAsFactors = F)

synteny <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/HN1_YMR_2nd_assemble_alignment_final_chr.csv",
                      sep=",",header = F,stringsAsFactors = F) 
gff_HN1 <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_HN1_wgdi/HN1.gff",stringsAsFactors = FALSE)


c=0
b=0
synteny_result_sub1 <- c()
synteny_result_sub2 <- c()

for (i in 1:177) {
  
  block_1_gene <- block1_count(singleton,all_block,i,10)
  block_2_gene <- block2_count(singleton,all_block,i,10)
  
  if(length(block_1_gene) > length(block_2_gene)){
    
    synteny_result_sub1 <- c(synteny_result_sub1,block_1_gene)
    synteny_result_sub2 <- c(synteny_result_sub2,block_2_gene)
  }else if(length(block_1_gene) < length(block_2_gene)){
    synteny_result_sub1 <- c(synteny_result_sub1,block_2_gene)
    synteny_result_sub2 <- c(synteny_result_sub2,block_1_gene)
  }else{
    block_1_gene <- block1_count(singleton,all_block,i,20)
    block_2_gene <- block2_count(singleton,all_block,i,20)
    
    if(length(block_1_gene) > length(block_2_gene)){
      block_1_gene <- block1_count(singleton,all_block,i,20)
      block_2_gene <- block2_count(singleton,all_block,i,20)
      synteny_result_sub1 <- c(synteny_result_sub1,block_1_gene)
      synteny_result_sub2 <- c(synteny_result_sub2,block_2_gene)
    }else if(length(block_1_gene) < length(block_2_gene)){
      block_1_gene <- block1_count(singleton,all_block,i,20)
      block_2_gene <- block2_count(singleton,all_block,i,20)
      synteny_result_sub1 <- c(synteny_result_sub1,block_2_gene)
      synteny_result_sub2 <- c(synteny_result_sub2,block_1_gene)
    }else{
      print(all_block[i,1])
    }
    
    
  }
  
}
singleton <- singleton[!singleton[,2] %in% c(synteny_result_sub1,synteny_result_sub2),]
synteny_result_sub1 <- c(synteny_result_sub1,singleton[,2])
intersect_data <- intersect(synteny_result_sub1,synteny_result_sub2)

synteny_result_sub1 <- synteny_result_sub1[!synteny_result_sub1 %in% intersect_data]
synteny_result_sub2 <- synteny_result_sub2[!synteny_result_sub2 %in% intersect_data]

sum(dim(synteny[synteny[,2] %in% synteny_result_sub1,])[1],
    dim(synteny[synteny[,3] %in% synteny_result_sub2,])[1])/
  sum(dim(synteny[synteny[,2] %in% synteny_result_sub1,])[1],
      dim(synteny[synteny[,3] %in% synteny_result_sub1,])[1],
      dim(synteny[synteny[,3] %in% synteny_result_sub2,])[1],
      dim(synteny[synteny[,2] %in% synteny_result_sub2,])[1])

## [1] 0.7654119

synteny_group1_sub1 <- synteny[synteny[,2] %in% synteny_result_sub1 ,]
synteny_group1_sub1<- synteny_group1_sub1[,c(1,2,4)]
synteny_group1_sub1[,3]<- "SG1"
colnames(synteny_group1_sub1) <- c("YMR","HN1","synteny_modify_class")
synteny_group1_sub2 <- synteny[synteny[,3] %in% synteny_result_sub1 ,]
synteny_group1_sub2<- synteny_group1_sub2[,c(1,3,4)]
synteny_group1_sub2[,3]<- "SG2"
colnames(synteny_group1_sub2) <- c("YMR","HN1","synteny_modify_class")
synteny_result_sub1 <- rbind(synteny_group1_sub1,synteny_group1_sub2)

synteny_group2_sub1 <- synteny[synteny[,2] %in% synteny_result_sub2 ,]
synteny_group2_sub1<- synteny_group2_sub1[,c(1,2,4)]
synteny_group2_sub1[,3]<- "SG1"
colnames(synteny_group2_sub1) <- c("YMR","HN1","synteny_modify_class")
synteny_group2_sub2 <- synteny[synteny[,3] %in% synteny_result_sub2 ,]
synteny_group2_sub2<- synteny_group2_sub2[,c(1,3,4)]
synteny_group2_sub2[,3]<- "SG2"
colnames(synteny_group2_sub2) <- c("YMR","HN1","synteny_modify_class")
synteny_result_sub2 <- rbind(synteny_group2_sub1,synteny_group2_sub2)


gff_HN1 <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi_subphaser/subphaser_gene.gff",stringsAsFactors = FALSE)

synteny_result_sub1[,4] <- NA
colnames(synteny_result_sub1)[c(3,4)] <- c("synteny_modify_class","orign_k_mer_class")

synteny_result_sub2[,4] <- NA
colnames(synteny_result_sub2)[c(3,4)] <- c("synteny_modify_class","orign_k_mer_class")


for (i in 1:dim(synteny_result_sub1)[1]) {
  synteny_result_sub1[i,4] <- gff_HN1[gff_HN1[,2] == synteny_result_sub1[i,2],][,8]
}

for (i in 1:dim(synteny_result_sub2)[1]) {
  synteny_result_sub2[i,4] <- gff_HN1[gff_HN1[,2] == synteny_result_sub2[i,2],][,8]
}


intersect(synteny_result_sub1[,2],synteny_result_sub2[,2])

write.table(synteny_result_sub1,"D:/002cns_evolution_1st_revision/001subgenome/HN1_LM_MF_PC/method2/synteny_result_sub1.xls"
            ,sep = "\t",col.names = T,row.names = F,quote = F)
write.table(synteny_result_sub2,"D:/002cns_evolution_1st_revision/001subgenome/HN1_LM_MF_PC/method2/synteny_result_sub2.xls"
            ,sep = "\t",col.names = T,row.names = F,quote = F)

```

