# 1. Identification of two subgenomes in P. somniferum

## 1.1 Blastp

```shell
## HN1 means P. somniferum
## YMR means P. rhoeas
blastp -query HN1.pep -db HN1 -evalue 1e-5 -outfmt 6 -num_threads 48 -max_target_seqs 11 -out HN1_11.blast
blastp -query YMR.pep -db YMR -evalue 1e-5 -outfmt 6 -num_threads 48 -max_target_seqs 11  -out YMR_11.blast
blastp -query HN1.pep -db YMR -evalue 1e-5  -outfmt 6 -num_threads 48 -max_target_seqs 10 -out HN1_YMR_10.blast
```



## 1.2 Filter Blastp results

```R
rm(list = ls())
gene_blast <- read.table("HN1_11.blast",
                       sep="\t",stringsAsFactors =FALSE)
gene_blast <- gene_blast[gene_blast[,12] >= 100,]
write.table(gene_blast,"Ps.blast"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```

```R
rm(list = ls())
gene_blast <- read.table("YMR_11.blast",
                       sep="\t",stringsAsFactors =FALSE)
gene_blast <- gene_blast[gene_blast[,12] >= 100,]
write.table(gene_blast,"Pr.blast"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```

## 1.3 Identify Tandem duplicates

```R
##make specific gff files for MCScanX
rm(list = ls())
gene_bed <- read.table("Papaver_somniferum.gene.bed",
                       sep="\t",stringsAsFactors =FALSE)

gene_bed[1:5,]
'''
    V1     V2     V3            V4 V5
1 chr1  59647  63768 Pso01G00010.0  +
2 chr1  62504  73904 Pso01G00020.0  -
3 chr1 142842 146495 Pso01G00030.0  +
4 chr1 145932 151557 Pso01G00040.0  -
5 chr1 151746 155004 Pso01G00050.0  +
'''
gene_bed[,1] <- paste0("Ps",gene_bed[,1])
gene_bed <- gene_bed[,c(1,4,2,3)]

write.table(gene_bed,"/subgenome/TD/Ps/Ps.gff"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

rm(list = ls())
gene_bed <- read.table("Papaver_rhoeas.gene.bed",
                       sep="\t",stringsAsFactors =FALSE)
gene_bed[,1] <- paste0("Pr",gene_bed[,1])
gene_bed <- gene_bed[,c(1,4,2,3)]

write.table(gene_bed,"/subgenome/TD/Pr/Pr.gff"
            ,sep = "\t",col.names = F,row.names = F,quote = F)
```



```shell
## Ps means P. somniferum
## Pr means P. rhoeas

./MCScanX ../../subgenome/TD/Ps/Ps
./MCScanX ../../subgenome/TD/Pr/Pr
```



## 1.4 Filter Blastp results (Remove TDs)

```R
rm(list = ls())
gene_blast <- read.table("HN1_YMR_10.blast",
                       sep="\t",stringsAsFactors =FALSE)
HN1_tandem <- read.table("/subgenome/TD/Ps/Ps.tandem",
                         sep=",",stringsAsFactors =FALSE)
HN1_tandem_list <- as.character(unique(c(HN1_tandem[,1],HN1_tandem[,2])))


YMR_tandem <- read.table("/subgenome/TD/Pr/Pr.tandem",
                         sep=",",stringsAsFactors =FALSE)
YMR_tandem_list <- as.character(unique(c(YMR_tandem[,1],YMR_tandem[,2])))

gene_blast<- gene_blast[!gene_blast[,1] %in% HN1_tandem_list,]
gene_blast<- gene_blast[!gene_blast[,2] %in% YMR_tandem_list,]

write.table(gene_blast,"/subgenome/HN1_YMR_wgdi/HN1_YMR_10_rdup.blast"
            ,sep = "\t",col.names = F,row.names = F,quote = F)

```

## 1.5 Identify collinearity and calculate selection pressure

```shell
[collinearity]
gff1 =  HN1.gff
gff2 =  YMR.gff
lens1 = HN1.lens
lens2 = YMR.lens
blast = HN1_YMR_10_rdup.blast
blast_reverse = false
multiple  = 2
process = 24
evalue = 1e-5
score = 100
grading = 50,40,25
mg = 40,40
pvalue = 0.2
repeat_number = 5
positon = order
savefile = HN1_YMR.collinearity_repeat5.txt

[ks]
cds_file =      HN1_YMR.cds.fa
pep_file =      HN1_YMR.pep
align_software = muscle
pairs_file = HN1_YMR.collinearity_repeat5.txt
ks_file = HN1_YMR.collinearity_repeat5.ks


[blockinfo]
blast = HN1_YMR_10_rdup.blast
gff1 =  HN1.gff
gff2 =  YMR.gff
lens1 = HN1.lens
lens2 = YMR.lens
collinearity = HN1_YMR.collinearity_repeat5.txt
score = 100
evalue = 1e-5
repeat_number = 5
position = order
ks = HN1_YMR.collinearity_repeat5.ks
ks_col = ks_YN00
savefile = HN1_YMR_block_information.csv

wgdi -bi HN1_YMR_wgdi.conf

[blockks]
lens1 = ../HN1_chr_only.lens
lens2 = ../YMR_chr_only.lens
genome1_name =  ../Papaver_somniferum
genome2_name =  ../Papaver_rhoeas
blockinfo = HN1_YMR_block_information.csv
pvalue = 0.2
tandem = true
tandem_length = 20
markersize = 1
area = -1,3
block_length =  10
figsize = 12,14.21
savefig = HN1_YMR_ks_dotplot.pdf

wgdi -bk  HN1_YMR_wgdi.conf


[correspondence]
blockinfo =  HN1_YMR_block_information.csv
lens1 = ../HN1_1.lens
lens2 = ../YMR_1.lens
tandem = true
tandem_length = 20
pvalue = 0.2
block_length = 10
multiple  = 2
homo = 0,1
savefile = HN1_YMR_new_block_information.csv 

wgdi -c HN1_YMR_wgdi.conf



[kspeaks]
blockinfo = HN1_YMR_new_block_information.csv 
pvalue = 0.2
tandem = true
block_length = 10
ks_area = 0,10
multiple  = 2
homo = 0.5,1
fontsize = 9
area = 0,3
figsize = 10,6.18
savefig = HN1_YMR.ks_median.distri.pdf
savefile = HN1_YMR.ks_median.distri.csv

wgdi -kp  HN1_YMR_wgdi.conf
```

  

## 1.6 Reconstruct subgenomes of P. somniferum (chromosome 6 in P. rhoeas as an example)

```R
## pre-defined functions

YMR_gene_list<-
  function(YMR_chr1_block_copy1){
    copy_1_YMRgene <- rbind()
    for (i in 1:dim(YMR_chr1_block_copy1)[1]) {
      syn_block1 <-  YMR_chr1_block_copy1[i,c("block2")]
      chr <- YMR_chr1_block_copy1[i,3]
      gene_order <- unlist(
        lapply(syn_block1, strsplit, split="_", fixed=TRUE, useBytes=TRUE)
      )
      
      gene_order <- unique(as.numeric(gene_order))
      syn_block1_gene <- as.data.frame(gene_order)
      syn_block1_gene[,2] <- chr
      colnames(syn_block1_gene) <- c("gene_order","chr")
      syn_block1_gene <- syn_block1_gene[,c(2,1)]
      copy_1_YMRgene <- rbind(copy_1_YMRgene,syn_block1_gene)
    }
    copy_1_YMRgene <- unique(copy_1_YMRgene)
    copy_1_YMRgene <- copy_1_YMRgene[,2]
    return(copy_1_YMRgene)
  }

HN1_gene_list<-
  function(df){
    gff_HN1 <- read.table("HN1.gff",stringsAsFactors = FALSE)
    
    gene_HN1 <- rbind()
    for (i in 1:dim(df)[1]) {
      syn_block1 <-  df[i,c("block1")]
      chr <- df[i,2]
      gene_order <- unlist(
        lapply(syn_block1, strsplit, split="_", fixed=TRUE, useBytes=TRUE)
      )
      
      gene_order <- unique(as.numeric(gene_order))
      gff_1 <- gff_HN1[gff_HN1[,1] == chr,]
      
      syn_block1_gene <- gff_1[ gff_1$V6 %in% gene_order, "V2"]
      syn_block1_gene <- as.data.frame(syn_block1_gene)
      syn_block1_gene[,2] <- df[i,1]
      colnames(syn_block1_gene) <- c("gene","block")
      gene_HN1 <- rbind(gene_HN1,syn_block1_gene)
    }
    return(as.character(unique(gene_HN1[,1])))
  }

first_assemble <-
  function(YMR_chr1_block_filter,copy_1_YMRgene,copy_1_HN1gene,copy_2_YMRgene,copy_2_HN1gene,wgd_pairs){
    for (filter in 1:dim(YMR_chr1_block_filter)[1]) {
      gene_order <-  YMR_gene_list(YMR_chr1_block_filter[filter,])
      gene_order_HN1 <- HN1_gene_list(YMR_chr1_block_filter[filter,])
      copy1_overlap_length <- length(intersect(gene_order,copy_1_YMRgene))
      copy2_overlap_length <- length(intersect(gene_order,copy_2_YMRgene))
      
      if(copy1_overlap_length < copy2_overlap_length){
        YMR_chr1_block_copy1 <- rbind(YMR_chr1_block_copy1,all_block[all_block$id == YMR_chr1_block_filter[filter,1],])
        copy_1_YMRgene <- unique(c(copy_1_YMRgene,gene_order))
        copy_1_HN1gene <- unique(c(copy_1_HN1gene,gene_order_HN1))
        
      }else if(copy1_overlap_length > copy2_overlap_length){
        YMR_chr1_block_copy2 <- rbind(YMR_chr1_block_copy2,all_block[all_block$id == YMR_chr1_block_filter[filter,1],])
        copy_2_YMRgene <- unique(c(copy_2_YMRgene,gene_order))
        copy_2_HN1gene <- unique(c(copy_2_HN1gene,gene_order_HN1))
        
      }else if(copy1_overlap_length == copy2_overlap_length){
        wgd_pairs_order <- rbind(wgd_pairs[wgd_pairs$id1 %in% gene_order_HN1,],wgd_pairs[wgd_pairs$id2 %in% gene_order_HN1,])
        wgd_pairs_list <- unique(c(wgd_pairs_order$id1,wgd_pairs_order$id2))
        
        wgd_pairs_list <- setdiff(wgd_pairs_list,gene_order_HN1)
        copy1_HN1_overlap_length <- length(intersect(wgd_pairs_list,copy_1_HN1gene))
        copy2_HN1_overlap_length <- length(intersect(wgd_pairs_list,copy_2_HN1gene))
        
        if(copy1_HN1_overlap_length < copy2_HN1_overlap_length){
          YMR_chr1_block_copy1 <- rbind(YMR_chr1_block_copy1,all_block[all_block$id == YMR_chr1_block_filter[filter,1],])
          copy_1_YMRgene <- unique(c(copy_1_YMRgene,gene_order))
          copy_1_HN1gene <- unique(c(copy_1_HN1gene,gene_order_HN1))
        }else if(copy1_HN1_overlap_length > copy2_HN1_overlap_length){
          YMR_chr1_block_copy2 <- rbind(YMR_chr1_block_copy2,all_block[all_block$id == YMR_chr1_block_filter[filter,1],])
          copy_2_YMRgene <- unique(c(copy_2_YMRgene,gene_order))
          copy_2_HN1gene <- unique(c(copy_2_HN1gene,gene_order_HN1))
        }else{
          print(filter)
        }
        
      }
      
    }
    out <- list(YMR_chr1_block_copy1=YMR_chr1_block_copy1,YMR_chr1_block_copy2=YMR_chr1_block_copy2)
    return(out)
  }



second_assemble <-
  function(YMR_chr1_block_filter,YMR_chr1_block_copy1,YMR_chr1_block_copy2,wgd_pairs){
    copy_1_YMRgene <- YMR_gene_list(YMR_chr1_block_copy1)
    copy_1_HN1gene <- HN1_gene_list(YMR_chr1_block_copy1)
    copy_2_YMRgene <- YMR_gene_list(YMR_chr1_block_copy2)
    copy_2_HN1gene <- HN1_gene_list(YMR_chr1_block_copy2)
    for (filter in 1:dim(YMR_chr1_block_filter)[1]) {
      gene_order <-  YMR_gene_list(YMR_chr1_block_filter[filter,])
      gene_order_HN1 <- HN1_gene_list(YMR_chr1_block_filter[filter,])
      copy1_overlap_length <- length(intersect(gene_order,copy_1_YMRgene))
      copy2_overlap_length <- length(intersect(gene_order,copy_2_YMRgene))
      
      if(copy1_overlap_length < copy2_overlap_length){
        YMR_chr1_block_copy1 <- rbind(YMR_chr1_block_copy1,all_block[all_block$id == YMR_chr1_block_filter[filter,1],])
        copy_1_YMRgene <- unique(c(copy_1_YMRgene,gene_order))
        copy_1_HN1gene <- unique(c(copy_1_HN1gene,gene_order_HN1))
        
      }else if(copy1_overlap_length > copy2_overlap_length){
        YMR_chr1_block_copy2 <- rbind(YMR_chr1_block_copy2,all_block[all_block$id == YMR_chr1_block_filter[filter,1],])
        copy_2_YMRgene <- unique(c(copy_2_YMRgene,gene_order))
        copy_2_HN1gene <- unique(c(copy_2_HN1gene,gene_order_HN1))
        
      }else if(copy1_overlap_length == copy2_overlap_length){
        wgd_pairs_order <- rbind(wgd_pairs[wgd_pairs$id1 %in% gene_order_HN1,],wgd_pairs[wgd_pairs$id2 %in% gene_order_HN1,])
        wgd_pairs_list <- unique(c(wgd_pairs_order$id1,wgd_pairs_order$id2))

        wgd_pairs_list <- setdiff(wgd_pairs_list,gene_order_HN1)
        
        copy1_HN1_overlap_length <- length(intersect(wgd_pairs_list,copy_1_HN1gene))
        copy2_HN1_overlap_length <- length(intersect(wgd_pairs_list,copy_2_HN1gene))
        
        if(copy1_HN1_overlap_length < copy2_HN1_overlap_length){
          YMR_chr1_block_copy1 <- rbind(YMR_chr1_block_copy1,all_block[all_block$id == YMR_chr1_block_filter[filter,1],])
          copy_1_YMRgene <- unique(c(copy_1_YMRgene,gene_order))
          copy_1_HN1gene <- unique(c(copy_1_HN1gene,gene_order_HN1))
        }else if(copy1_HN1_overlap_length > copy2_HN1_overlap_length){
          YMR_chr1_block_copy2 <- rbind(YMR_chr1_block_copy2,all_block[all_block$id == YMR_chr1_block_filter[filter,1],])
          copy_2_YMRgene <- unique(c(copy_2_YMRgene,gene_order))
          copy_2_HN1gene <- unique(c(copy_2_HN1gene,gene_order_HN1))
        }else{
          print(filter)
        }
        
      }
      
    }
    out <- list(YMR_chr1_block_copy1=YMR_chr1_block_copy1,YMR_chr1_block_copy2=YMR_chr1_block_copy2)
    return(out)
  }


third_assemble <-
  function(YMR_chr1_block_filter,YMR_chr1_block_copy1,YMR_chr1_block_copy2,wgd_pairs){
    gff_HN1 <- read.table("HN1.gff",stringsAsFactors = FALSE)
    result <- rbind()
    copy_1_YMRgene <- YMR_gene_list(YMR_chr1_block_copy1)
    copy_1_HN1gene <- HN1_gene_list(YMR_chr1_block_copy1)
    copy_2_YMRgene <- YMR_gene_list(YMR_chr1_block_copy2)
    copy_2_HN1gene <- HN1_gene_list(YMR_chr1_block_copy2)
    for (filter in 1:dim(YMR_chr1_block_filter)[1]) {
      gene_order <-  YMR_gene_list(YMR_chr1_block_filter[filter,])
      gene_order_HN1 <- HN1_gene_list(YMR_chr1_block_filter[filter,])
      
      wgd_pairs_order <- rbind(wgd_pairs[wgd_pairs$id1 %in% gene_order_HN1,],wgd_pairs[wgd_pairs$id2 %in% gene_order_HN1,])
      wgd_pairs_list <- unique(c(wgd_pairs_order$id1,wgd_pairs_order$id2))
      wgd_pairs_list <- setdiff(wgd_pairs_list,gene_order_HN1)
      
      gff_HN1_1 <- gff_HN1[gff_HN1[,2] %in% wgd_pairs_list,]
      if(dim(gff_HN1_1)[1] != 0){
        gff_HN1_1[,7] <- filter
        result <- rbind(result,gff_HN1_1)
      }
      
    }
    return(result)
  }

```



```R
all_block <- read.csv("HN1_YMR.ks_median.distri.csv",stringsAsFactors = F)
## select recent WGD related blocks
all_block <- all_block[all_block$ks_median < 0.32,]
YMR_chr1_block <- all_block[all_block$chr2 == "chr6",]
## chr4 and chr5 in P. somniferum are two important scaffolds for chr6 in P. rhoeas
## chr4 to copy1 and chr5 to copy2
YMR_chr1_block_copy1 <- YMR_chr1_block[YMR_chr1_block$chr1 == "chr4",]
YMR_chr1_block_copy2 <- YMR_chr1_block[YMR_chr1_block$chr1 == "chr5",]
YMR_chr1_block_filter <- YMR_chr1_block[YMR_chr1_block$chr1 != "chr4" & YMR_chr1_block$chr1 != "chr5",]
##sort
YMR_chr1_block_filter <- YMR_chr1_block_filter[order(YMR_chr1_block_filter$length,decreasing = T),]


copy_1_YMRgene <- YMR_gene_list(YMR_chr1_block_copy1)
copy_1_HN1gene <- HN1_gene_list(YMR_chr1_block_copy1)
copy_2_YMRgene <- YMR_gene_list(YMR_chr1_block_copy2)
copy_2_HN1gene <- HN1_gene_list(YMR_chr1_block_copy2)

wgd_pairs <- read.table("WGD.ks",stringsAsFactors = F,sep="\t",header = T)
wgd_pairs <- wgd_pairs[wgd_pairs$ks_YN00 <0.32,]

assemble <- first_assemble(YMR_chr1_block_filter,copy_1_YMRgene,copy_1_HN1gene,copy_2_YMRgene,copy_2_HN1gene,wgd_pairs)

YMR_chr1_block_copy1 <- assemble$YMR_chr1_block_copy1
YMR_chr1_block_copy2 <- assemble$YMR_chr1_block_copy2

gff_YMR <- read.table("YMR.gff",stringsAsFactors = FALSE)
copy_1_YMRgene <- YMR_gene_list(YMR_chr1_block_copy1)
copy_2_YMRgene <- YMR_gene_list(YMR_chr1_block_copy2)


write.table(copy_1_YMRgene,"subgenome2_YMR.csv",sep = ",",col.names = T,row.names = F,quote = F)
write.table(copy_2_YMRgene,"subgenome1_YMR.csv",sep = ",",col.names = T,row.names = F,quote = F)
write.table(YMR_chr1_block_copy1,"subgenome2.csv",sep = ",",col.names = T,row.names = F,quote = F)
write.table(YMR_chr1_block_copy2,"subgenome1.csv",sep = ",",col.names = T,row.names = F,quote = F)

```

