# 1. ATAC-seq data analysis



## 1.1 The overlapped ratios of ACRs

### 1.1.1 Get coordinates of 1kb upstream for gene

```R

for(kb_i in c(1000)){
  
  HN1_gff <- read.csv("/HN1.gff",stringsAsFactors = F,sep="\t",header = F)
  
  gene_distance_cal <- function(a){
    kb = kb_i
    if(a[5] == "+"){
      return(paste0(a[1],",",as.numeric(a[3])-kb+1,",",as.numeric(a[3]),",",a[7],",",a[5]))
    }else{
      return(paste0(a[1],",",as.numeric(a[4]),",",as.numeric(a[4])+kb-1,",",a[7],",",a[5]))
    }
  }
  
  gene_distance_1kb <- as.data.frame(apply(HN1_gff, 1, gene_distance_cal))
  colnames(gene_distance_1kb) <- "V1"
  library(tidyverse)
  gene_distance_1kb <- separate(data = gene_distance_1kb, 
                                col = V1, 
                                into = c("chr","1kb_start","1kb_end","gene","strand"),
                                sep = ",")
  
  gene_distance_1kb[,2] <- as.numeric(gene_distance_1kb[,2])-1
  gene_distance_1kb[,3] <- as.numeric(gene_distance_1kb[,3])
  
  for (i in 1:55316) {
    if(gene_distance_1kb[i,2] < 0){
      gene_distance_1kb[i,2] = 0
    }
  } 
  
  write.table(gene_distance_1kb,paste0("/gene_upstream_",kb_i,"_all.bed"),
              sep = "\t",col.names = F,row.names = F,quote = F)
  rm(list=ls())
  
}


```
### 1.1.2 Get coordinates of overlapped CNSs


```shell
##HN1_mask_YMR_mask.bed is a file containing the coordinates of the CNSs
bedtools intersect  -a gene_upstream_1000_all.bed -b ./HN1_mask_YMR_mask.bed -f 0.015 > cns_overlap_1kb.bed
```

```R
## modify the coordinates of CNSs

for(kb_i in 1:1){
  cns_overlap <- read.csv(paste0("/cns_overlap_",kb_i,"kb.bed"),
                          stringsAsFactors = F,sep="\t",header = F)
  cns_overlap <- cns_overlap[order(cns_overlap$V1,cns_overlap$V2,cns_overlap$V3),]
  cns_overlap <- cns_overlap[,c(-4,-5)]
  write.table(cns_overlap,paste0("/cns_overlap_",kb_i,"kb_modify.bed"),
              sep = "\t",col.names = F,row.names = F,quote = F)
}
```

### 1.1.3 Get coordinates of overlapped Non_CNSs

```shell
bedtools subtract  -a gene_upstream_1000_all.bed -b cns_overlap_1kb_modify.bed > non_cns_1kb.bed
```

```R
## modify the coordinates of non_CNSs

for(kb_i in 1:1){
  cns_overlap <- read.csv(paste0("/non_cns_",kb_i,"kb.bed"),
                          stringsAsFactors = F,sep="\t",header = F)
  cns_overlap <- cns_overlap[order(cns_overlap$V1,cns_overlap$V2,cns_overlap$V3),]
  cns_overlap <- cns_overlap[,c(-4,-5)]
  write.table(cns_overlap,paste0("/non_cns_",kb_i,"kb_modify.bed"),
              sep = "\t",col.names = F,row.names = F,quote = F)
}
```



### 1.1.4 Count ATAC reads overlapped with CNSs or Non_CNSs

```shell
cd /ATAC/coordinates 

ls *_modify.bed| while read id;do(multiBamSummary BED-file  --BED $id --bamfiles /ATAC/ATAC_sort_bam_filter_sort_1/*.bam  --numberOfProcessors 48 -out /ATAC/reads_count_coordinates/$id.npz --outRawCounts /ATAC/reads_count_coordinates/$id.tab);done

```

### 1.1.5 Plot Figure 2B

```R
atac_cns <- read.csv("/reads_count_coordinates/cns_overlap_1kb_modify.bed.tab",sep = "\t",stringsAsFactors = F)
atac_cns[,1] <- atac_cns[,3]-atac_cns[,2]
atac_cns[,4:21] <-  atac_cns[,4:21]/atac_cns[,1]


atac_cns_5_tissues <- as.data.frame(matrix(0,nrow = 50434,ncol = 5))
colnames(atac_cns)
colnames(atac_cns_5_tissues) <- c("petal","stem","fine_root","leaf","tap_root")
for (i in 1:5) {
  atac_cns_5_tissues[,i] <- atac_cns[,3*(i-1)+7]+atac_cns[,3*(i-1)+8]+atac_cns[,3*(i-1)+9]
}
atac_cns_5_tissues[,6] <- "cns"

atac_non_cns <- read.csv("/reads_count_coordinates/non_cns_1kb_modify.bed.tab",sep = "\t",stringsAsFactors = F)
atac_non_cns[,1] <- atac_non_cns[,3]-atac_non_cns[,2]
atac_non_cns <- atac_non_cns[atac_non_cns[,1] >= 15,]

atac_non_cns[,4:21] <-  atac_non_cns[,4:21]/atac_non_cns[,1]


atac_non_cns_5_tissues <- as.data.frame(matrix(0,nrow = 84985,ncol = 5))
colnames(atac_non_cns)
colnames(atac_non_cns_5_tissues) <- c("petal","stem","fine_root","leaf","tap_root")
for (i in 1:5) {
  atac_non_cns_5_tissues[,i] <- atac_non_cns[,3*(i-1)+7]+atac_non_cns[,3*(i-1)+8]+atac_non_cns[,3*(i-1)+9]
}
atac_non_cns_5_tissues[,6] <- "non_cns"

plot_data <- rbind()
for (i in 1:5) {
  data_cns <- atac_cns_5_tissues[,c(i,6)]
  data_cns[,2] <- paste0(colnames(atac_cns_5_tissues)[i],"_cns")
  colnames(data_cns) <- c("tissue","type")
  data_cns[,3] <- paste0(colnames(atac_cns_5_tissues)[i])
  
  
  data_non_cns <- atac_non_cns_5_tissues[,c(i,6)]
  data_non_cns[,2] <- paste0(colnames(atac_non_cns_5_tissues)[i],"_non_cns")
  colnames(data_non_cns) <- c("tissue","type")
  data_non_cns[,3] <- paste0(colnames(atac_non_cns_5_tissues)[i])
  
  
  plot_data <- rbind(plot_data,data_cns)
  plot_data <- rbind(plot_data,data_non_cns)
}
fontstyle="Arial"
bins=50
pointsize=2
scale=100
xymax=100000
fontsize=15
linesize=1


library(ggplot2)


petal <-
  ggplot(plot_data[plot_data[,3]=="petal",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = " ",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,0.8),
                       labels = c(seq(0,1,0.1)))

stem <-
  ggplot(plot_data[plot_data[,3]=="stem",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = " ",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,1),
                       labels = c(seq(0,1,0.1)))



fine_root <-
  ggplot(plot_data[plot_data[,3]=="fine_root",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = " ",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,0.4),
                       labels = c(seq(0,1,0.1)))


leaf <-
  ggplot(plot_data[plot_data[,3]=="leaf",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = "",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,0.4),
                       labels = c(seq(0,1,0.1)))



tap_root <-
  ggplot(plot_data[plot_data[,3]=="tap_root",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = " ",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,0.8),
                       labels = c(seq(0,1,0.1)))


```



