# 1. ATAC-seq data analysis

## 1.1 call peak

```
## bam files were got from
Xu, Y., Jia, Y.Y., Wang, B., Guo, L., Guo, M.Y., Che, X.F. and Ye, K. (2022b) The tissue-specific chromatin accessibility landscape of Papaver somniferum. bioRxiv, 2022.2004.2013.487984.
```

```python
## merge_atac_bam.py

CLASS_INDEX = {"stem","fine_root","petal","leaf","tap_root"}

rule all:
    input:
        expand("ATAC_seq_{tissue_class}_merge.bam",tissue_class=CLASS_INDEX),
        expand("ATAC_seq_{tissue_class}_merge.bam.bai",tissue_class=CLASS_INDEX)


rule samtools_merge:
    input:
        "ATAC_seq_{tissue_class}_1_rmdup_filter.bam",
        "ATAC_seq_{tissue_class}_2_rmdup_filter.bam",
        "ATAC_seq_{tissue_class}_3_rmdup_filter.bam"
                
    output:
        "ATAC_seq_{tissue_class}_merge.bam"
    threads: 20
    shell:
        "samtools merge -@ 20 {output} {input[0]} {input[1]} {input[2]}"

rule samtools_merge_index:
    input:
        "ATAC_seq_{tissue_class}_merge.bam"
    output:
        "ATAC_seq_{tissue_class}_merge.bam.bai"
    shell:
        "samtools index -@ 20 {input} {output}"

```



```R
c <- file( "/data/home/xuyu/001atac/ATAC_sort_bam_filter_sort_1_merge_bam/macs2/macs2_q_0.05_command_new", "w" )
for (i in list.files("/data/home/xuyu/001atac/ATAC_sort_bam_filter_sort_1_merge_bam/",pattern="*.bam$" )) {
  writeLines( paste0("macs2 callpeak -t ",i," -f BAMPE -n ./macs2_q_0.05_new/",i," -q 0.05 -g 2408287349 --keep-dup all"), c )
}
close( c ) 

```

```shell
#!/bin/bash
#PBS -N macs2_q_0.05_merge_new
#PBS -o /data/home/xuyu/001atac/pbs/macs2_q_0.05_merge_new.stdout
#PBS -e /data/home/xuyu/001atac/pbs/macs2_q_0.05_merge_new.stderr
#PBS -q batch
#PBS -l nodes=1:ppn=48
#PBS -l mem=350G

export PATH=$PATH:/data/home/xuyu/anaconda3/envs/macs2/bin/

cat ./macs2/macs2_q_0.05_command_new| parallel -j 5
```



## 1.2 Figure 2B

### 1.2.1 Get coordinates of 1kb upstream for gene

```R
for(kb_i in c(1000)){
  
  HN1_gff <- read.csv("D:/002cns_evolution/motif/HN1.gff",stringsAsFactors = F,sep="\t",header = F)
  
  gene_distance_cal <- function(a){
    kb = kb_i
    if(a[5] == "+"){
      return(paste0(a[1],",",as.numeric(a[3])-kb+1,",",as.numeric(a[3]),",",a[7],",",a[5]))
    }else{
      return(paste0(a[1],",",as.numeric(a[4]),",",as.numeric(a[4])+kb-1,",",a[7],",",a[5]))
    }
  }
  
  gene_distance_1kb <- as.data.frame(apply(HN1_gff, 1, gene_distance_cal))
  colnames(gene_distance_1kb) <- "V1"
  library(tidyverse)
  gene_distance_1kb <- separate(data = gene_distance_1kb, 
                                col = V1, 
                                into = c("chr","1kb_start","1kb_end","gene","strand"),
                                sep = ",")
  
  gene_distance_1kb[,2] <- as.numeric(gene_distance_1kb[,2])-1
  gene_distance_1kb[,3] <- as.numeric(gene_distance_1kb[,3])
  
  for (i in 1:55316) {
    if(gene_distance_1kb[i,2] < 0){
      gene_distance_1kb[i,2] = 0
    }
  } 
  
  write.table(gene_distance_1kb,paste0("D:/002cns_evolution_1st_revision/002CNS_1_ATAC/real_promoter/gene_upstream_",kb_i,"_all.bed"),
              sep = "\t",col.names = F,row.names = F,quote = F)
  rm(list=ls())
  
}

```
### 1.2.2 Get coordinates of overlapped CNSs


```shell
##HN1_mask_YMR_mask.bed is a file containing the coordinates of the CNSs
bedtools intersect  -a gene_upstream_1000_all.bed -b ../HN1_mask_YMR_mask.bed > cns_overlap_1kb.bed
```

```R
## modify the coordinates of CNSs

for(kb_i in 1:1){
  cns_overlap <- read.csv(paste0("/cns_overlap_",kb_i,"kb.bed"),
                          stringsAsFactors = F,sep="\t",header = F)
  cns_overlap <- cns_overlap[order(cns_overlap$V1,cns_overlap$V2,cns_overlap$V3),]
  cns_overlap <- cns_overlap[,c(-4,-5)]
  write.table(cns_overlap,paste0("/cns_overlap_",kb_i,"kb_modify.bed"),
              sep = "\t",col.names = F,row.names = F,quote = F)
}
```

### 1.2.3 Get coordinates of overlapped Non_CNSs

```shell
bedtools subtract  -a gene_upstream_1000_all.bed -b cns_overlap_1kb_modify.bed > non_cns_1kb.bed
```

```R
## modify the coordinates of non_CNSs

for(kb_i in 1:1){
  cns_overlap <- read.csv(paste0("/non_cns_",kb_i,"kb.bed"),
                          stringsAsFactors = F,sep="\t",header = F)
  cns_overlap <- cns_overlap[order(cns_overlap$V1,cns_overlap$V2,cns_overlap$V3),]
  cns_overlap <- cns_overlap[,c(-4,-5)]
  write.table(cns_overlap,paste0("/non_cns_",kb_i,"kb_modify.bed"),
              sep = "\t",col.names = F,row.names = F,quote = F)
}
```



### 1.2.4 Count ATAC reads overlapped with CNSs or Non_CNSs

```shell
cd /ATAC/coordinates 

ls *_modify.bed| while read id;do(multiBamSummary BED-file  --BED $id --bamfiles /ATAC/ATAC_sort_bam_filter_sort_1/*.bam  --numberOfProcessors 48 -out /ATAC/reads_count_coordinates/$id.npz --outRawCounts /ATAC/reads_count_coordinates/$id.tab);done

```

### 1.2.5 Plot Figure 2B

```R
atac_cns <- read.csv("D:/002cns_evolution_1st_revision/002CNS_1_ATAC/real_promoter/reads_count_coordinates/cns_overlap_1kb_modify.bed.tab",sep = "\t",stringsAsFactors = F)
atac_cns[,1] <- atac_cns[,3]-atac_cns[,2]
atac_cns <- atac_cns[atac_cns[,1] >= 30,]
atac_cns[,4:21] <-  atac_cns[,4:21]/atac_cns[,1]


atac_cns_5_tissues <- as.data.frame(matrix(0,nrow = dim(atac_cns)[1],ncol = 5))
colnames(atac_cns)
colnames(atac_cns_5_tissues) <- c("h_petal","j_stem","x_fine_root","y_leaf","z_tap_root")
for (i in 1:5) {
  atac_cns_5_tissues[,i] <- atac_cns[,3*(i-1)+7]+atac_cns[,3*(i-1)+8]+atac_cns[,3*(i-1)+9]
}
atac_cns_5_tissues[,6] <- "cns"

atac_non_cns <- read.csv("D:/002cns_evolution_1st_revision/002CNS_1_ATAC/real_promoter/reads_count_coordinates/non_cns_1kb_modify.bed.tab",sep = "\t",stringsAsFactors = F)
atac_non_cns[,1] <- atac_non_cns[,3]-atac_non_cns[,2]
atac_non_cns <- atac_non_cns[atac_non_cns[,1] >= 30,]

atac_non_cns[,4:21] <-  atac_non_cns[,4:21]/atac_non_cns[,1]


atac_non_cns_5_tissues <- as.data.frame(matrix(0,nrow = dim(atac_non_cns)[1],ncol = 5))
colnames(atac_non_cns)
colnames(atac_non_cns_5_tissues) <- c("h_petal","j_stem","x_fine_root","y_leaf","z_tap_root")
for (i in 1:5) {
  atac_non_cns_5_tissues[,i] <- atac_non_cns[,3*(i-1)+7]+atac_non_cns[,3*(i-1)+8]+atac_non_cns[,3*(i-1)+9]
}
atac_non_cns_5_tissues[,6] <- "non_cns"

plot_data <- rbind()
for (i in 1:5) {
  data_cns <- atac_cns_5_tissues[,c(i,6)]
  data_cns[,2] <- paste0(colnames(atac_cns_5_tissues)[i],"_cns")
  colnames(data_cns) <- c("tissue","type")
  data_cns[,3] <- paste0(colnames(atac_cns_5_tissues)[i])
  
  
  data_non_cns <- atac_non_cns_5_tissues[,c(i,6)]
  data_non_cns[,2] <- paste0(colnames(atac_non_cns_5_tissues)[i],"_non_cns")
  colnames(data_non_cns) <- c("tissue","type")
  data_non_cns[,3] <- paste0(colnames(atac_non_cns_5_tissues)[i])
  
  
  plot_data <- rbind(plot_data,data_cns)
  plot_data <- rbind(plot_data,data_non_cns)
}
fontstyle="Arial"
bins=50
pointsize=2
scale=100
xymax=100000
fontsize=15
linesize=1


library(ggplot2)
library(ggpubr)
library(ggbreak)


compaired <- list(c("h_petal_cns","h_petal_non_cns"),
                  c("j_stem_cns","j_stem_non_cns"),
                  c("y_leaf_cns","y_leaf_non_cns"),
                  c("z_tap_root_cns","z_tap_root_non_cns"),
                  c("x_fine_root_cns","x_fine_root_non_cns"))
ggplot(plot_data, aes(x=type,y=tissue))+
  geom_boxplot(aes(fill=type),width=0.5,outlier.shape = NA)+ 
  geom_signif(comparisons = compaired,
              step_increase = 0,
              map_signif_level = F,
              test = wilcox.test)+
  labs(y = " ",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )



h_petal <-
  ggplot(plot_data[plot_data[,3]=="h_petal",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = " ",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,0.8),
                       labels = c(seq(0,1,0.1)))

j_stem <-
  ggplot(plot_data[plot_data[,3]=="j_stem",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = " ",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,1),
                       labels = c(seq(0,1,0.1)))



x_fine_root <-
  ggplot(plot_data[plot_data[,3]=="x_fine_root",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = " ",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,0.4),
                       labels = c(seq(0,1,0.1)))


y_leaf <-
  ggplot(plot_data[plot_data[,3]=="y_leaf",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = "",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,0.4),
                       labels = c(seq(0,1,0.1)))



z_tap_root <-
  ggplot(plot_data[plot_data[,3]=="z_tap_root",], aes(x=type,y=tissue))+
  geom_boxplot(fill=c('#E69F00','#999999'),width=0.5,outlier.shape = NA)+ 
  
  labs(y = " ",x="")+
  theme(
    axis.text.x=element_text( angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    axis.text.y=element_text(angle=0,hjust=0.5,vjust=0.5,size=fontsize,colour = "black"),
    plot.title=element_text(size=fontsize+3,hjust = 0.5,colour = "black"),
    axis.title.x = element_text(face = "plain",size=fontsize+1,vjust=-1,colour = "black"),
    axis.title.y = element_text(face = "plain",size=fontsize+1,vjust=2.5,colour = "black"),
    panel.background = element_rect(fill = "white", colour = "black"),
    panel.border=element_rect(fill = NA, color="black"),
    panel.grid.major.x=element_blank(),
    panel.grid.minor.x=element_blank(),
    panel.grid.major.y=element_blank(), #
    panel.grid.minor.y=element_blank(),
    axis.line.x=element_line(colour="black"),
    axis.line.y=element_line(colour="black")
    
  )+scale_y_continuous(breaks = c(seq(0,1,0.1)),limits = c(0,0.8),
                       labels = c(seq(0,1,0.1)))

## 5 10
library(ggpubr)
library(gridExtra)
ggarrange(arrangeGrob(h_petal),
          arrangeGrob(j_stem),
          arrangeGrob(x_fine_root),
          arrangeGrob(y_leaf),
          arrangeGrob(z_tap_root),
          nrow = 1) 

```



## 1.3 Intron (Figure S4)

```R
library("GenomicFeatures")
gtf <- makeTxDbFromGFF("D:/002cns_evolution/dCNS_comparative/CNS/比较/HN1.gff3")
introndata <- intronsByTranscript(gtf,use.names=TRUE)
introndata <- unlist(introndata)

introndata_data <- as.data.frame(introndata@ranges)
library(tidyverse)
introndata_data <- separate(introndata_data,col="names",into = c("gene","names_1","names2"),
                            sep = "-")
introndata_data <- introndata_data[,-c(5,6)]

gff_HN1 <- read.table("D:/002cns_evolution_1st_revision/001subgenome/HN1_YMR_wgdi/HN1.gff",stringsAsFactors = FALSE)
gff_HN1 <- gff_HN1[,c(1,2)]


introndata_data_merge <- merge(introndata_data,gff_HN1,by.x="gene",by.y="V2")
introndata_data_merge <- introndata_data_merge[,c(5,2,3)]

##transfer to bed file , the start minus 1
introndata_data_merge[,2] <- introndata_data_merge[,2]-1
introndata_data_merge <- unique(introndata_data_merge)

write.table(introndata_data_merge,"D:/002cns_evolution_1st_revision/002CNS_1_ATAC/real_intron/intron_all.bed",
            sep = "\t",col.names = F,row.names = F,quote = F)

```

```shell
bedtools intersect  -a intron_all.bed -b ../HN1_mask_YMR_mask.bed  > cns_overlap_intron.bed
```

```R
cns_overlap <- read.csv(paste0("D:/002cns_evolution_1st_revision/002CNS_1_ATAC/real_intron/cns_overlap_intron.bed"),
                        stringsAsFactors = F,sep="\t",header = F)
cns_overlap <- cns_overlap[order(cns_overlap$V1,cns_overlap$V2,cns_overlap$V3),]
cns_overlap <- unique(cns_overlap)
write.table(cns_overlap,paste0("D:/002cns_evolution_1st_revision/002CNS_1_ATAC/real_intron/cns_overlap_intron_modify.bed"),
            sep = "\t",col.names = F,row.names = F,quote = F)
```

```shell
bedtools subtract  -a intron_all.bed -b cns_overlap_intron_modify.bed > non_cns_intron.bed
```

```R

cns_overlap <- read.csv(paste0("D:/002cns_evolution_1st_revision/002CNS_1_ATAC/real_intron/non_cns_intron.bed"),
                        stringsAsFactors = F,sep="\t",header = F)
cns_overlap <- cns_overlap[order(cns_overlap$V1,cns_overlap$V2,cns_overlap$V3),]
cns_overlap <- unique(cns_overlap)
write.table(cns_overlap,paste0("D:/002cns_evolution_1st_revision/002CNS_1_ATAC/real_intron/non_cns_intron_modify.bed"),
            sep = "\t",col.names = F,row.names = F,quote = F)

```

```shell

ls *_modify.bed| while read id;do(multiBamSummary BED-file  --BED $id --bamfiles /ATAC/ATAC_sort_bam_filter_sort_1/*.bam  --numberOfProcessors 48 -out /ATAC/reads_count_coordinates/$id.npz --outRawCounts /ATAC/reads_count_coordinates/$id.tab);done

```

## 1.4 Figure S5

```shell
cd /002cns_evolution_1st_revision/002CNS_1_ATAC/overlap_with_ATAC_peak/upstream_1kb/

## peak get from 1.1
## YE MEANS LEAF; HUA MEANS PETAL;JING MEANS STEM; ZHUGEN MEANS TAP ROOT;XUGEN MEANS FINE ROOT

bedtools intersect -a ../ATAC_seq_YE_merge.bam_peaks.narrowPeak -b ../gene_upstream_1kb_all.bed  > YE.bed
bedtools intersect -a ../ATAC_seq_HUA_merge.bam_peaks.narrowPeak -b ../gene_upstream_1kb_all.bed  > HUA.bed
bedtools intersect -a ../ATAC_seq_JING_merge.bam_peaks.narrowPeak -b ../gene_upstream_1kb_all.bed  > JING.bed
bedtools intersect -a ../ATAC_seq_ZHUGEN_merge.bam_peaks.narrowPeak -b ../gene_upstream_1kb_all.bed  > ZHUGEN.bed
bedtools intersect -a ../ATAC_seq_XUGEN_merge.bam_peaks.narrowPeak -b ../gene_upstream_1kb_all.bed  > XUGEN.bed


bedtools intersect -a YE.bed -b ../cns_overlap_1kb_modify.bed  > YE_CNS_peak.bed
bedtools intersect -a HUA.bed -b ../cns_overlap_1kb_modify.bed  > HUA_CNS_peak.bed
bedtools intersect -a JING.bed -b ../cns_overlap_1kb_modify.bed > JING_CNS_peak.bed
bedtools intersect -a ZHUGEN.bed -b ../cns_overlap_1kb_modify.bed > ZHUGEN_CNS_peak.bed
bedtools intersect -a XUGEN.bed -b ../cns_overlap_1kb_modify.bed > XUGEN_CNS_peak.bed


```

```R
result_data <- as.data.frame(matrix(NA,ncol = 2,nrow = 5))
i=1
for (tissue in c("HUA","JING","YE","XUGEN","ZHUGEN")) {
  
  data_all <- read.table(paste0("D:/002cns_evolution_1st_revision/002CNS_1_ATAC/overlap_with_ATAC_peak/upstream_1kb/",tissue,".bed"),
                         sep = "\t",stringsAsFactors = F)
  data_all[,11] <- data_all[,3]-data_all[,2]
  
  
  data <- read.table(paste0("D:/002cns_evolution_1st_revision/002CNS_1_ATAC/overlap_with_ATAC_peak/upstream_1kb/",tissue,"_CNS_peak.bed"),
                     sep = "\t",stringsAsFactors = F)
  data[,11] <- data[,3]-data[,2]
  
  result_data[i,1] <- tissue
  result_data[i,2] <- sum(data[,11])/sum(data_all[,11])
  i <- i+1
}



write.csv(result_data,"D:/002cns_evolution_1st_revision/002CNS_1_ATAC/overlap_with_ATAC_peak/upstream_1kb/count.csv")

```

