# 1. Motif analysis

## 1.1 Get coordinates of 1kb upstream for gene

```
## HN1 means P. somniferum
## YMR means P. rhoeas
First, coordinates of upstream 1kb regions for both_conservation and both_specialization related genes were extracted.
```

```R
HN1_gff <- read.csv("D:/002cns_evolution/motif/HN1.gff",stringsAsFactors = F,sep="\t",header = F)

kb_i = 1000

gene_distance_cal <- function(a){
  kb = kb_i
  if(a[5] == "+"){
    return(paste0(a[1],",",as.numeric(a[3])-kb+1,",",as.numeric(a[3]),",",a[7],",",a[5]))
  }else{
    return(paste0(a[1],",",as.numeric(a[4]),",",as.numeric(a[4])+kb-1,",",a[7],",",a[5]))
  }
}

gene_distance_1kb <- as.data.frame(apply(HN1_gff, 1, gene_distance_cal))
colnames(gene_distance_1kb) <- "V1"
library(tidyverse)
gene_distance_1kb <- separate(data = gene_distance_1kb, 
                              col = V1, 
                              into = c("chr","1kb_start","1kb_end","gene","strand"),
                              sep = ",")

gene_distance_1kb[,2] <- as.numeric(gene_distance_1kb[,2])-1
gene_distance_1kb[,3] <- as.numeric(gene_distance_1kb[,3])

for (i in 1:55316) {
  if(gene_distance_1kb[i,2] < 0){
    gene_distance_1kb[i,2] = 0
  }
} 


evolution_method = "pearson"
quantile_normalize = "quantile"
threshold_method = "mean"

both_retain <- read.csv(paste0("D:/002cns_evolution_1st_revision/003_1transcriptom_evo/CNS_all_different_methdos_k_mer/",evolution_method,"_",
                               quantile_normalize,"_",threshold_method,"_TPM/COR/both_retain_ka_ks.csv"),
                        stringsAsFactors = F,header = T)
both_retain_con <- both_retain[both_retain[,7] == "Both_Conservation",]

gene_distance_1kb_dup_con <- gene_distance_1kb[gene_distance_1kb$gene %in% c(both_retain_con[,2],both_retain_con[,3]),]


both_retain_spe <- both_retain[both_retain[,7] == "Both_Specialization",]


gene_distance_1kb_dup_spe <- gene_distance_1kb[gene_distance_1kb$gene %in% c(both_retain_spe[,2],both_retain_spe[,3]),]



write.table(gene_distance_1kb_dup_con,"D:/002cns_evolution_1st_revision/004_1motif_dup_con_spe/conv/gene_upstream_1kb_all.bed",
            sep = "\t",col.names = F,row.names = F,quote = F)

write.table(gene_distance_1kb_dup_spe,"D:/002cns_evolution_1st_revision/004_1motif_dup_con_spe/spe/gene_upstream_1kb_all.bed",
            sep = "\t",col.names = F,row.names = F,quote = F)


```

## 1.2  Get coordinates of overlapped CNSs

```shell

cd conv
bedtools intersect  -a gene_upstream_1kb_all.bed -b ../HN1_mask_YMR_mask.bed -f 0.015 > cns_overlap_1kb.bed

cd ../spe
bedtools intersect  -a gene_upstream_1kb_all.bed -b ../HN1_mask_YMR_mask.bed -f 0.015> cns_overlap_1kb.bed
```

```R

## modify the coordinates
cns_overlap <- read.csv("D:/002cns_evolution_1st_revision/004_1motif_dup_con_spe/conv/cns_overlap_1kb.bed",stringsAsFactors = F,sep="\t",header = F)
cns_overlap <- cns_overlap[order(cns_overlap$V1,cns_overlap$V2,cns_overlap$V3),]
cns_overlap <- unique(cns_overlap)
cns_overlap[,5] <- 1:dim(cns_overlap)[1]
cns_overlap[,4] <- paste0(cns_overlap[,4],"_",cns_overlap[,5])
cns_overlap <- cns_overlap[,-5]
cns_overlap <- unique(cns_overlap)
write.table(cns_overlap,"D:/002cns_evolution_1st_revision/004_1motif_dup_con_spe/conv/cns_overlap_1kb_modify.bed",
            sep = "\t",col.names = F,row.names = F,quote = F)

cns_overlap <- read.csv("D:/002cns_evolution_1st_revision/004_1motif_dup_con_spe/spe/cns_overlap_1kb.bed",stringsAsFactors = F,sep="\t",header = F)
cns_overlap <- cns_overlap[order(cns_overlap$V1,cns_overlap$V2,cns_overlap$V3),]
cns_overlap <- unique(cns_overlap)

cns_overlap[,5] <- 1:dim(cns_overlap)[1]
cns_overlap[,4] <- paste0(cns_overlap[,4],"_",cns_overlap[,5])
cns_overlap <- cns_overlap[,-5]
cns_overlap <- unique(cns_overlap)
write.table(cns_overlap,"D:/002cns_evolution_1st_revision/004_1motif_dup_con_spe/spe/cns_overlap_1kb_modify.bed",
            sep = "\t",col.names = F,row.names = F,quote = F)


```

## 1.3 Shuffle 10000 times for CNSs annotations

```shell


cd /data/home/xuyu/002cns_evolution_1st_revision/004_1motif_dup_con_spe/conv

mkdir shuffle

for((i=1;i<=10000;i++));
do
bedtools shuffle -i cns_overlap_1kb_modify.bed -g ../HN1.GENOME -incl gene_upstream_1kb_all.bed -seed $i  > ./shuffle/shuffle_$i.bed
done


cd /data/home/xuyu/002cns_evolution_1st_revision/004_1motif_dup_con_spe/spe
mkdir shuffle
for((i=1;i<=10000;i++));
do
bedtools shuffle -i cns_overlap_1kb_modify.bed -g ../HN1.GENOME -incl gene_upstream_1kb_all.bed -seed $i  > ./shuffle/shuffle_$i.bed
done
```



## 1.4 Find motifs

```shell
cd /data/home/xuyu/002cns_evolution_1st_revision/004_1motif_dup_con_spe/spe/shuffle

ls | awk '{print("annotatePeaks.pl ./shuffle/"$1" /data/home/xuyu/genome_info/Papaver_somniferum.ref.fa -m ../arab.dap.motifs > ./shuffle_motif/"$1".txt")}' > ../command_all

cd /data/home/xuyu/002cns_evolution_1st_revision/004_1motif_dup_con_spe/conv/shuffle

ls | awk '{print("annotatePeaks.pl ./shuffle/"$1" /data/home/xuyu/genome_info/Papaver_somniferum.ref.fa -m ../arab.dap.motifs > ./shuffle_motif/"$1".txt")}' > ../command_all
```



## 1.5 Calculate the number of each motifs in CNSs or permutation CNSs

### 1.5.1 both_conservation

```R
library("optparse")
library(tidyverse)
library(stringr)
options(stringsAsFactors = FALSE)

option_list <- list(
    make_option( c("-c", "--cutoff"), type = "integer", default = 0, help = "select_threshold ")
) 

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)
x1 = opt$cutoff

conv_motif_summary_result <-
  function(start_i){
    
    motif_binding <- read.table(paste0("conv","/shuffle_motif/shuffle_",start_i,".bed.txt"),sep="\t",stringsAsFactors = F,header = T)
    motif_binding <- motif_binding[,-c(2:21)]
    motif_binding <- as.data.frame(t(motif_binding))
    colnames(motif_binding) <- motif_binding[1,]
    motif_binding <- motif_binding[-1,]
    
    motif_binding[,dim(motif_binding)[2]+1] <- rownames(motif_binding)
    
    motif_binding_filter <- separate(data = motif_binding, 
                                     col = colnames(motif_binding)[dim(motif_binding)[2]],
                                     into = c("V1","V2"), sep = "\\.")
    
    rownames(motif_binding_filter) <- motif_binding_filter[,dim(motif_binding_filter)[2]-1]
    motif_binding_filter <- motif_binding_filter[,-c(dim(motif_binding_filter)[2],dim(motif_binding_filter)[2]-1)]
    
    
    cal_kuohao<- function(a){
      return(str_count(a, "\\("))
    }
    
    motif_binding_filter_1 <- apply(motif_binding_filter[1:dim(motif_binding_filter)[1],], 1, cal_kuohao)
    motif_binding_filter_1 <- as.data.frame(t(motif_binding_filter_1))
    
    write.table(motif_binding_filter_1,paste0("conv","/motif_statistics/shuffle_",start_i,".txt"),
                sep = "\t",col.names = T,row.names = T,quote = F)
    
    cal_summary<- function(b){
      return(sum(b))
    }
    motif_binding_filter_summary <- as.data.frame(apply(motif_binding_filter_1[1:dim(motif_binding_filter_1)[1],], 1, cal_summary))
    colnames(motif_binding_filter_summary) <- "summary"
    write.table(motif_binding_filter_summary,paste0("conv","/motif_summary/shuffle_",start_i,".txt"),
                sep = "\t",col.names = T,row.names = T,quote = F)
    
  }
conv_motif_summary_result(x1)
```





### 1.5.2 both_specialization

```R
library("optparse")
library(tidyverse)
library(stringr)
options(stringsAsFactors = FALSE)

option_list <- list(
    make_option( c("-c", "--cutoff"), type = "integer", default = 0, help = "select_threshold")
) 

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)
x1 = opt$cutoff

spe_motif_summary_result <-
  function(start_i){
    
    motif_binding <- read.table(paste0("spe","/shuffle_motif/shuffle_",start_i,".bed.txt"),sep="\t",stringsAsFactors = F,header = T)
    motif_binding <- motif_binding[,-c(2:21)]
    motif_binding <- as.data.frame(t(motif_binding))
    colnames(motif_binding) <- motif_binding[1,]
    motif_binding <- motif_binding[-1,]
    
    motif_binding[,dim(motif_binding)[2]+1] <- rownames(motif_binding)
    
    motif_binding_filter <- separate(data = motif_binding, 
                                     col = colnames(motif_binding)[dim(motif_binding)[2]],
                                     into = c("V1","V2"), sep = "\\.")
    
    rownames(motif_binding_filter) <- motif_binding_filter[,dim(motif_binding_filter)[2]-1]
    motif_binding_filter <- motif_binding_filter[,-c(dim(motif_binding_filter)[2],dim(motif_binding_filter)[2]-1)]
    
    
    cal_kuohao<- function(a){
      return(str_count(a, "\\("))
    }
    
    motif_binding_filter_1 <- apply(motif_binding_filter[1:dim(motif_binding_filter)[1],], 1, cal_kuohao)
    motif_binding_filter_1 <- as.data.frame(t(motif_binding_filter_1))
    
    write.table(motif_binding_filter_1,paste0("spe","/motif_statistics/shuffle_",start_i,".txt"),
                sep = "\t",col.names = T,row.names = T,quote = F)
    
    cal_summary<- function(b){
      return(sum(b))
    }
    motif_binding_filter_summary <- as.data.frame(apply(motif_binding_filter_1[1:dim(motif_binding_filter_1)[1],], 1, cal_summary))
    colnames(motif_binding_filter_summary) <- "summary"
    write.table(motif_binding_filter_summary,paste0("spe","/motif_summary/shuffle_",start_i,".txt"),
                sep = "\t",col.names = T,row.names = T,quote = F)
    
  }
spe_motif_summary_result(x1)
```



## 1.6 Motifs annotations

```R
library(UniprotR)

list <- read.csv("motif.csv",stringsAsFactors = F)
colnames(list) <- c("tf","uniprot_name")
a <- GetProteinFunction(list[,2])
a <- a[,c(1,10)]
a[,1] <- rownames(a)
colnames(a) <- c("uniprot_name","general_function_description")
b <- GetProteinGOInfo(list[,2])
b[,1] <- rownames(b)
colnames(b) <- c("uniprot_name","GO_ALL","GO_BP","GO_MF","GO_CC")
tf_anno <- merge(a,b,by="uniprot_name")

list_merge <- merge(list,tf_anno,by="uniprot_name")


write.csv(list_merge,"motif_annotation.csv")

```

